{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "# Quick background - the evasin family\n",
    "\n",
    "Evasins are a family of secreted proteins found in ticks.\n",
    "\n",
    "They bind to cytokines and modulate the host immune response.\n",
    "\n",
    "The two well studied evasins, Evasin-1 and Evasin-4 from *Rhipicephalus sanguineus* are residues 114 and 127 residues long respectively.\n",
    "\n",
    "The family contains a conserved patterns of four disulpide bonds.\n",
    "\n",
    "The three-dimensional structure of Evasin-1 from *Rhipicephalus sanguineus* is known (PDB ID: `3FPR`).\n",
    "\n",
    "Just for fun, let's take a look."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: nglview in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages\n",
      "Requirement already satisfied: ipywidgets==7 in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from nglview)\n",
      "Requirement already satisfied: numpy in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from nglview)\n",
      "Requirement already satisfied: ipython>=4.0.0; python_version >= \"3.3\" in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from ipywidgets==7->nglview)\n",
      "Requirement already satisfied: traitlets>=4.3.1 in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from ipywidgets==7->nglview)\n",
      "Requirement already satisfied: widgetsnbextension~=3.0.0 in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from ipywidgets==7->nglview)\n",
      "Requirement already satisfied: nbformat>=4.2.0 in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from ipywidgets==7->nglview)\n",
      "Requirement already satisfied: ipykernel>=4.5.1 in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from ipywidgets==7->nglview)\n",
      "Requirement already satisfied: setuptools>=18.5 in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from ipython>=4.0.0; python_version >= \"3.3\"->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: simplegeneric>0.8 in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from ipython>=4.0.0; python_version >= \"3.3\"->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: appnope; sys_platform == \"darwin\" in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from ipython>=4.0.0; python_version >= \"3.3\"->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: jedi>=0.10 in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from ipython>=4.0.0; python_version >= \"3.3\"->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: decorator in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from ipython>=4.0.0; python_version >= \"3.3\"->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: pickleshare in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from ipython>=4.0.0; python_version >= \"3.3\"->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: pygments in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from ipython>=4.0.0; python_version >= \"3.3\"->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: pexpect; sys_platform != \"win32\" in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from ipython>=4.0.0; python_version >= \"3.3\"->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: prompt-toolkit<2.0.0,>=1.0.4 in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from ipython>=4.0.0; python_version >= \"3.3\"->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: ipython-genutils in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from traitlets>=4.3.1->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: six in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from traitlets>=4.3.1->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: notebook>=4.4.1 in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from widgetsnbextension~=3.0.0->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: jupyter-core in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from nbformat>=4.2.0->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: jsonschema!=2.5.0,>=2.4 in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from nbformat>=4.2.0->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: tornado>=4.0 in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from ipykernel>=4.5.1->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: jupyter-client in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from ipykernel>=4.5.1->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: parso==0.1.0 in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from jedi>=0.10->ipython>=4.0.0; python_version >= \"3.3\"->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from pexpect; sys_platform != \"win32\"->ipython>=4.0.0; python_version >= \"3.3\"->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: wcwidth in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from prompt-toolkit<2.0.0,>=1.0.4->ipython>=4.0.0; python_version >= \"3.3\"->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: terminado>=0.3.3; sys_platform != \"win32\" in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from notebook>=4.4.1->widgetsnbextension~=3.0.0->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: nbconvert in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from notebook>=4.4.1->widgetsnbextension~=3.0.0->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: jinja2 in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from notebook>=4.4.1->widgetsnbextension~=3.0.0->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: python-dateutil>=2.1 in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from jupyter-client->ipykernel>=4.5.1->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: pyzmq>=13 in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from jupyter-client->ipykernel>=4.5.1->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: entrypoints>=0.2.2 in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from nbconvert->notebook>=4.4.1->widgetsnbextension~=3.0.0->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: bleach in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from nbconvert->notebook>=4.4.1->widgetsnbextension~=3.0.0->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: mistune>=0.7.4 in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from nbconvert->notebook>=4.4.1->widgetsnbextension~=3.0.0->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: pandocfilters>=1.4.1 in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from nbconvert->notebook>=4.4.1->widgetsnbextension~=3.0.0->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: testpath in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from nbconvert->notebook>=4.4.1->widgetsnbextension~=3.0.0->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: MarkupSafe>=0.23 in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from jinja2->notebook>=4.4.1->widgetsnbextension~=3.0.0->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: html5lib!=1.0b1,!=1.0b2,!=1.0b3,!=1.0b4,!=1.0b5,!=1.0b6,!=1.0b7,!=1.0b8,>=0.99999999pre in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from bleach->nbconvert->notebook>=4.4.1->widgetsnbextension~=3.0.0->ipywidgets==7->nglview)\n",
      "Requirement already satisfied: webencodings in /Users/perry/.virtualenvs/jupyter_tag_filter_exporter/lib/python3.6/site-packages (from html5lib!=1.0b1,!=1.0b2,!=1.0b3,!=1.0b4,!=1.0b5,!=1.0b6,!=1.0b7,!=1.0b8,>=0.99999999pre->bleach->nbconvert->notebook>=4.4.1->widgetsnbextension~=3.0.0->ipywidgets==7->nglview)\n"
     ]
    }
   ],
   "source": [
    "# If we start a line with `!`, Jupyter runs a Unix shell command\n",
    "# This is a special Jupyter feature, not regular Python\n",
    "!pip install nglview"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f01c364b2445407bb990db2c18cd10a6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "A Jupyter Widget"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import urllib\n",
    "import nglview as nv\n",
    "\n",
    "pdb_id = '3FPR'\n",
    "\n",
    "# We can download the coordinates from the PDB \n",
    "# and save them to a local file like this\n",
    "pdb_fn = \"%s.pdb\" % pdb_id\n",
    "urllib.request.urlretrieve(\"https://files.rcsb.org/download/%s\" % pdb_fn, filename=pdb_fn)\n",
    "\n",
    "view = nv.show_file('3FPR.pdb')\n",
    "\n",
    "view.representations = []\n",
    "\n",
    "# Show the bonds in the cysteines as thick cylinders\n",
    "view.add_licorice('CYS')\n",
    "# Show the backbone as a cartoon, coloured by secondary structure type\n",
    "view.add_cartoon(color='sstruc')\n",
    "\n",
    "# Hint: You can rotate by holding the left mouse button and zoom with the scroll wheel.\n",
    "view.display()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Running external tools from within Python - a BLAST search\n",
    "\n",
    "Let's run a BLAST search against the locally installed Uniprot database to find protein sequences similar to Evasin-1 and Evasin-4.\n",
    "\n",
    "`evasins_canonical.fasta` contains the sequences for Evasin-1 and Evasin-4. We know from prior work that they are paralogous, so we will BLAST search them both and combine all the hits.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [],
   "source": [
    "import subprocess\n",
    "database = '/mnt/references/uniprot/2017_10/uniprot_2017_10'\n",
    "\n",
    "# Since this takes ~5 min or more to run, we've prepared the \"evasins_canonical_uniprot.blast7\" output file earlier.\n",
    "# Please don't run it, if everyone does you'll probably bring down the Jupyter server !\n",
    "\n",
    "# process = subprocess.run('blastp -query evasins_canonical.fasta -out evasins_canonical_uniprot.blast7 -outfmt 7 -evalue 10 -db %s' % database, \n",
    "#                          shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "\n",
    "# print(process)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "BLAST+ format 7 (`-outfmt 7`) looks like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "slideshow": {
     "slide_type": ""
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# BLASTP 2.2.31+\r\n",
      "# Query: sp|P0C8E7|EVA1_RHISA Evasin-1 OS=Rhipicephalus sanguineus PE=1 SV=1\r\n",
      "# Database: /mnt/references/uniprot/2017_10/uniprot_2017_10\r\n",
      "# Fields: query id, subject id, % identity, alignment length, mismatches, gap opens, q. start, q. end, s. start, s. end, evalue, bit score\r\n",
      "# 123 hits found\r\n",
      "sp|P0C8E7|EVA1_RHISA\tsp|P0C8E7|EVA1_RHISA\t100.00\t114\t0\t0\t1\t114\t1\t114\t9e-78\t237\r\n",
      "sp|P0C8E7|EVA1_RHISA\ttr|C9W1N8|C9W1N8_RHISA\t99.12\t114\t1\t0\t1\t114\t1\t114\t3e-77\t235\r\n",
      "sp|P0C8E7|EVA1_RHISA\ttr|A0A131Z694|A0A131Z694_RHIAP\t57.01\t107\t41\t2\t13\t114\t19\t125\t5e-34\t125\r\n",
      "sp|P0C8E7|EVA1_RHISA\ttr|A0A131ZAV1|A0A131ZAV1_RHIAP\t64.44\t90\t31\t1\t23\t111\t18\t107\t6e-34\t125\r\n",
      "sp|P0C8E7|EVA1_RHISA\ttr|A0A131X8U1|A0A131X8U1_9ACAR\t56.18\t89\t38\t1\t25\t113\t1\t88\t3e-30\t115\r\n"
     ]
    }
   ],
   "source": [
    "!head evasins_canonical_uniprot.blast7"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "The first few lines (`#`) contain metadata about the search that we can skip for now.\n",
    "Each BLAST hit is a single line with tab-seperated fields. The second field is the sequence ID we want.\n",
    "We will grab all the hits irrespective of e-value.\n",
    "\n",
    "First we will parse the `evasins_canonical_uniprot.blast7` file to create a list of hits.\n",
    "\n",
    "Once we have the IDs of each hit, we can use `blastdbcmd` to extract the full length sequences from the BLAST database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['sp|P0C8E7|EVA1_RHISA',\n",
       " 'tr|C9W1N8|C9W1N8_RHISA',\n",
       " 'tr|A0A131Z694|A0A131Z694_RHIAP',\n",
       " 'tr|A0A131ZAV1|A0A131ZAV1_RHIAP',\n",
       " 'tr|A0A131X8U1|A0A131X8U1_9ACAR']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blast_hits = []\n",
    "with open('evasins_canonical_uniprot.blast7', 'r') as fh:\n",
    "    for line in fh:\n",
    "        if line[0] != '#':\n",
    "            blast_hits.append(line.split('\\t')[1])\n",
    "\n",
    "blast_hits[0:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": [
     "challenge"
    ]
   },
   "source": [
    "## Challenge\n",
    "\n",
    "Rather than extracting the ID of the hits, can you also instead parse the E-value field and create a list of E-values ?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": [
     "solution"
    ]
   },
   "source": [
    "## Solution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": [
     "solution"
    ]
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['9e-78', '3e-77', '5e-34', '6e-34', '3e-30']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "e_values = []\n",
    "with open('evasins_canonical_uniprot.blast7', 'r') as fh:\n",
    "    for line in fh:\n",
    "        if line[0] != '#':\n",
    "            # hit_id = line.split('\\t')[1]\n",
    "            e = line.split('\\t')[10]\n",
    "            e_values.append(e)\n",
    "\n",
    "e_values[0:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "**Protip**: The scikit-bio package has a BLAST parser that captures all the fields into a Pandas dataframe. The BioPython package can also parse BLAST XML output. If I wasn't trying to demonstrate basic file parsing, I'd probably use one of those packages instead.\n",
    "\n",
    "http://scikit-bio.org/docs/latest/generated/skbio.io.format.blast7.html#module-skbio.io.format.blast7"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "We want to ensure there are no duplicate IDs in our list (since the `blastp` command was supplied with two query sequences, Evasin-1 and Evasin-4, the `evasins_canonical_uniprot.blast7` output will contain the results of two BLAST searches).\n",
    "\n",
    "If we turn the `list` into a `set`, Python will automatically discard duplicate IDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of BLAST hits: 149\n"
     ]
    }
   ],
   "source": [
    "print(\"Total number of BLAST hits:\", len(blast_hits))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "blast_hits = set(blast_hits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of hits with duplicates removed: 132\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of hits with duplicates removed:\", len(blast_hits))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "Now we use `blastdbcmd` with out list of unique IDs to extract the full FASTA format sequence from the BLAST database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'blastdbcmd -db /mnt/references/uniprot/2017_10/uniprot_2017_10 -entry \"tr|A0A224YE88|A0A224YE88_9ACAR,tr|A0A131YT06|A0A131YT06_RHIAP,tr|A0A023GD07|A0A023GD07_9ACAR,tr|A0A131X968|A0A131X968_9ACAR,tr|A0A023FSY4|A0A023FSY4_9ACAR,tr|A0A224Y3D0|A0A224Y3D0_9ACAR,tr|A0A131Z694|A0A131Z694_RHIAP,tr|A0A023FC49|A0A023FC49_9ACAR,tr|A0A023FC02|A0A023FC02_9ACAR,tr|A0A131YRX0|A0A131YRX0_RHIAP,tr|G3MIX6|G3MIX6_9ACAR,tr|A0A131Z6N2|A0A131Z6N2_RHIAP,tr|C9W1C3|C9W1C3_RHISA,tr|A0A223FZ22|A0A223FZ22_RHIMP,tr|A0A023FSY8|A0A023FSY8_9ACAR,tr|A0A023FTK5|A0A023FTK5_9ACAR,tr|A0A023FSX6|A0A023FSX6_9ACAR,tr|A0A023FCX2|A0A023FCX2_9ACAR,tr|A0A131YUQ8|A0A131YUQ8_RHIAP,tr|A0A023G2I5|A0A023G2I5_9ACAR,tr|C9W1Q9|C9W1Q9_RHISA,tr|A0A023G5I5|A0A023G5I5_9ACAR,tr|A0A023G2M2|A0A023G2M2_9ACAR,tr|A0A023G9R6|A0A023G9R6_9ACAR,tr|A0A131X8U1|A0A131X8U1_9ACAR,tr|A0A131YHW9|A0A131YHW9_RHIAP,tr|A0A131ZAV1|A0A131ZAV1_RHIAP,tr|A0A1E1WXB0|A0A1E1WXB0_9ACAR,tr|A0A023FQ59|A0A023FQ59_9ACAR,tr|A0A0C9S461|A0A0C9S461_AMBAM,tr|A0A023FC51|A0A023FC51_9ACAR,tr|A0A023FD15|A0A023FD15_9ACAR,tr|L7M8Z8|L7M8Z8_9ACAR,tr|A0A023G2G7|A0A023G2G7_9ACAR,tr|C9W1N8|C9W1N8_RHISA,tr|A0A023G506|A0A023G506_9ACAR,tr|A0A023FC58|A0A023FC58_9ACAR,tr|A0A1E1X165|A0A1E1X165_9ACAR,tr|A0A0C9R689|A0A0C9R689_AMBAM,tr|A0A1E5SLG8|A0A1E5SLG8_9BACT,tr|A0A023G6B6|A0A023G6B6_9ACAR,tr|G3MSY7|G3MSY7_9ACAR,tr|A0A023FZW1|A0A023FZW1_9ACAR,tr|A0A023FF05|A0A023FF05_9ACAR,tr|A0A0C9SDE2|A0A0C9SDE2_AMBAM,tr|A0A131XLS7|A0A131XLS7_9ACAR,tr|A0A023FG15|A0A023FG15_9ACAR,tr|A0A131YNV9|A0A131YNV9_RHIAP,tr|L7LT68|L7LT68_9ACAR,tr|A0A131YFL3|A0A131YFL3_RHIAP,tr|A0A023G4G4|A0A023G4G4_9ACAR,tr|G3MSY5|G3MSY5_9ACAR,tr|A0A023G159|A0A023G159_9ACAR,tr|G3MIM7|G3MIM7_9ACAR,tr|A0A023G8N3|A0A023G8N3_9ACAR,tr|L7MC83|L7MC83_9ACAR,tr|A0A1E1WWB2|A0A1E1WWB2_9ACAR,tr|A0A023FF01|A0A023FF01_9ACAR,tr|A0A224Y3C9|A0A224Y3C9_9ACAR,tr|A0A1E1X1C4|A0A1E1X1C4_9ACAR,tr|A0A023FC52|A0A023FC52_9ACAR,tr|A0A023FRC3|A0A023FRC3_9ACAR,tr|A0A023G9N2|A0A023G9N2_9ACAR,tr|A0A023FE33|A0A023FE33_9ACAR,tr|A0A0F3KQC8|A0A0F3KQC8_9NEIS,tr|A0A0C9R5W8|A0A0C9R5W8_AMBAM,tr|A0A224YCT0|A0A224YCT0_9ACAR,tr|A0A0C9SAC5|A0A0C9SAC5_AMBAM,sp|P0C8E9|EVA4_RHISA,tr|A0A023FEC6|A0A023FEC6_9ACAR,tr|G3MSY6|G3MSY6_9ACAR,tr|A0A0C9RX43|A0A0C9RX43_AMBAM,tr|A0A023FDY8|A0A023FDY8_9ACAR,tr|A0A023G4E4|A0A023G4E4_9ACAR,tr|A0A0C9S3J8|A0A0C9S3J8_AMBAM,tr|L7MC72|L7MC72_9ACAR,tr|A0A023FFB5|A0A023FFB5_9ACAR,tr|A0A023FZ70|A0A023FZ70_9ACAR,tr|A0A023GCB8|A0A023GCB8_9ACAR,tr|A0A131Z586|A0A131Z586_RHIAP,tr|A0A023FQC1|A0A023FQC1_9ACAR,tr|W6KVX4|W6KVX4_9TRYP,tr|A0A224YCD7|A0A224YCD7_9ACAR,tr|A0A224YE69|A0A224YE69_9ACAR,tr|A0A0C9S4J0|A0A0C9S4J0_AMBAM,tr|A0A131Z2K5|A0A131Z2K5_RHIAP,tr|A0A023G9N9|A0A023G9N9_9ACAR,tr|A0A023FDV6|A0A023FDV6_9ACAR,tr|A0A023G065|A0A023G065_9ACAR,tr|A0A0C9RVT5|A0A0C9RVT5_AMBAM,tr|A0A224Y913|A0A224Y913_9ACAR,tr|A0A023G2D8|A0A023G2D8_9ACAR,tr|A0A023G9N5|A0A023G9N5_9ACAR,tr|A0A023G9Q7|A0A023G9Q7_9ACAR,tr|A0A131YS46|A0A131YS46_RHIAP,tr|A0A023FCU0|A0A023FCU0_9ACAR,sp|P0C8E7|EVA1_RHISA,tr|A0A023FG84|A0A023FG84_9ACAR,tr|A0A023FBW7|A0A023FBW7_9ACAR,tr|A0A023FQ65|A0A023FQ65_9ACAR,tr|G3MJ83|G3MJ83_9ACAR,tr|C9W1Q7|C9W1Q7_RHISA,tr|A0A023FQ58|A0A023FQ58_9ACAR,tr|A0A023FT53|A0A023FT53_9ACAR,tr|A0A023GCC3|A0A023GCC3_9ACAR,tr|A0A023FFD0|A0A023FFD0_9ACAR,tr|A0A131YQC0|A0A131YQC0_RHIAP,tr|L7M8Y5|L7M8Y5_9ACAR,tr|A0A224Y3D8|A0A224Y3D8_9ACAR,tr|A0A023FCJ0|A0A023FCJ0_9ACAR,tr|A0A0C9R5A0|A0A0C9R5A0_AMBAM,tr|A0A023FQ55|A0A023FQ55_9ACAR,tr|A0A023GD11|A0A023GD11_9ACAR,tr|A0A131Z5M0|A0A131Z5M0_RHIAP,tr|A0A023FDQ9|A0A023FDQ9_9ACAR,tr|L7MC74|L7MC74_9ACAR,tr|A0A131Z6L7|A0A131Z6L7_RHIAP,tr|A0A023FED6|A0A023FED6_9ACAR,tr|A0A1E1XT85|A0A1E1XT85_9ACAR,tr|L7MAB0|L7MAB0_9ACAR,tr|A0A023FG08|A0A023FG08_9ACAR,tr|A0A131Z5R5|A0A131Z5R5_RHIAP,tr|A0A023G9R1|A0A023G9R1_9ACAR,tr|L7MA80|L7MA80_9ACAR,tr|A0A023FRB3|A0A023FRB3_9ACAR,tr|A0A023G109|A0A023G109_9ACAR,tr|A0A023FQ56|A0A023FQ56_9ACAR,tr|A0A224YE79|A0A224YE79_9ACAR,tr|A0A131XJ87|A0A131XJ87_9ACAR,tr|A0A224YCA7|A0A224YCA7_9ACAR,tr|A0A223FZ27|A0A223FZ27_RHIMP,tr|A0A1E1X0E4|A0A1E1X0E4_9ACAR\" >evasins_uniprot_hits.fasta'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We redirect to a file with '>' using the Unix shell command, rather than \n",
    "# capturing stdout in process.stdout\n",
    "cmd = 'blastdbcmd -db %s -entry \"%s\" >evasins_uniprot_hits.fasta' % (database, ','.join(blast_hits))\n",
    "cmd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\n",
    "process = subprocess.run(cmd, shell=True, stderr=subprocess.PIPE)\n",
    "\n",
    "# If you'd like to skip this step (eg if you don't have blastdbcmd or a local Uniprot BLAST database available),\n",
    "# you can run:\n",
    "#\n",
    "# !cp evasins_uniprot_hits.fasta.complete evasins_uniprot_hits.fasta\n",
    "#\n",
    "# to create a copy of the expected output sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 588,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">tr|A0A023G8N3|A0A023G8N3_9ACAR Putative evasin 1 OS=Amblyomma triste PE=2 SV=1\r\n",
      "MTTTTSATLIFLLYVSQLLVGFSRNEASDPPQTDEDCEYYDPAVDNITCTIQSLNTTGDPIPVGCLATCENSTRYLPNGT\r\n",
      "ECLGISQHVANRMQGNVNYTCPVGLCYRGVCQRNGLGIDCWHDTPPPNSTNVTTKAPTTLTSGRDL\r\n",
      ">tr|A0A023G9R1|A0A023G9R1_9ACAR Putative evasin 1 OS=Amblyomma triste PE=2 SV=1\r\n",
      "MTTTSAALIFLLYISQLLVVFSGNEVSDPPLIDEDCEYYDPSEDNITCTIRSLNTTGRPIPVGCLAMCENSTRRLHNGTE\r\n",
      "CLGISDKVANRMQGNVTYTCPVGLCYRGVCDRNGLGIDCWHNTPPPNSTNVTTNASTTSLPTSSRDL\r\n",
      ">tr|A0A023G4G4|A0A023G4G4_9ACAR Putative secreted protein OS=Amblyomma triste PE=2 SV=1\r\n",
      "MSSQSMFQLFLFVCGVAASVSANTDQSLLSCTSHKVLVMTNLGPFQVGCTQPCSNMKGMAAISEDGEECIDITRDGARKM\r\n",
      "PRRLEHRCPTGQCIKGTCKPDDLQVLCWYTGKDDATTPPRGT\r\n",
      ">tr|A0A023FCX2|A0A023FCX2_9ACAR Putative secreted mucin OS=Amblyomma cajennense PE=2 SV=1\r\n"
     ]
    }
   ],
   "source": [
    "# Take a peek at the output of our `blastdbcmd`, in the file evasin1_uniprot.fasta\n",
    "!head evasins_uniprot_hits.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Using the Python ecosystem: FASTA parsing with skbio.io\n",
    "\n",
    "http://scikit-bio.org/docs/latest/io.html\n",
    "\n",
    "Since you know how to parse a file now, you could go ahead and write a FASTA format parser to read in our new set of putative evasin sequences. However, for most common file formats a parser already exists to make your life easier. We are going to use the one `scikit-bio` provides (`Bio.SeqIO` from Biopython would be another reasonable option).\n",
    "\n",
    "*If you are curious, you'll find a quick-n-dirty example of a hand written FASTA parser in the `DIY_FASTA.ipynb` notebook.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 589,
   "metadata": {
    "scrolled": false,
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: scikit-bio==0.5.1 in /usr/local/lib/python3.5/dist-packages\n",
      "Requirement already satisfied: numpy>=1.9.2 in /usr/local/lib/python3.5/dist-packages (from scikit-bio==0.5.1)\n",
      "Requirement already satisfied: pandas>=0.18.0 in /usr/local/lib/python3.5/dist-packages (from scikit-bio==0.5.1)\n",
      "Requirement already satisfied: nose>=1.3.7 in /usr/local/lib/python3.5/dist-packages (from scikit-bio==0.5.1)\n",
      "Requirement already satisfied: natsort>=4.0.3 in /usr/local/lib/python3.5/dist-packages (from scikit-bio==0.5.1)\n",
      "Requirement already satisfied: scipy>=0.15.1 in /usr/local/lib/python3.5/dist-packages (from scikit-bio==0.5.1)\n",
      "Requirement already satisfied: lockfile>=0.10.2 in /usr/local/lib/python3.5/dist-packages (from scikit-bio==0.5.1)\n",
      "Requirement already satisfied: decorator>=3.4.2 in /usr/local/lib/python3.5/dist-packages (from scikit-bio==0.5.1)\n",
      "Requirement already satisfied: CacheControl>=0.11.5 in /usr/local/lib/python3.5/dist-packages (from scikit-bio==0.5.1)\n",
      "Requirement already satisfied: matplotlib>=1.4.3 in /usr/local/lib/python3.5/dist-packages (from scikit-bio==0.5.1)\n",
      "Requirement already satisfied: IPython>=3.2.0 in /usr/local/lib/python3.5/dist-packages (from scikit-bio==0.5.1)\n",
      "Requirement already satisfied: python-dateutil>=2 in /usr/local/lib/python3.5/dist-packages (from pandas>=0.18.0->scikit-bio==0.5.1)\n",
      "Requirement already satisfied: pytz>=2011k in /usr/local/lib/python3.5/dist-packages (from pandas>=0.18.0->scikit-bio==0.5.1)\n",
      "Requirement already satisfied: msgpack-python in /usr/local/lib/python3.5/dist-packages (from CacheControl>=0.11.5->scikit-bio==0.5.1)\n",
      "Requirement already satisfied: requests in /usr/local/lib/python3.5/dist-packages (from CacheControl>=0.11.5->scikit-bio==0.5.1)\n",
      "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.5/dist-packages (from matplotlib>=1.4.3->scikit-bio==0.5.1)\n",
      "Requirement already satisfied: six>=1.10 in /usr/local/lib/python3.5/dist-packages (from matplotlib>=1.4.3->scikit-bio==0.5.1)\n",
      "Requirement already satisfied: pyparsing!=2.0.0,!=2.0.4,!=2.1.2,!=2.1.6,>=1.5.6 in /usr/local/lib/python3.5/dist-packages (from matplotlib>=1.4.3->scikit-bio==0.5.1)\n",
      "Requirement already satisfied: setuptools>=18.5 in /usr/local/lib/python3.5/dist-packages (from IPython>=3.2.0->scikit-bio==0.5.1)\n",
      "Requirement already satisfied: traitlets>=4.2 in /usr/local/lib/python3.5/dist-packages (from IPython>=3.2.0->scikit-bio==0.5.1)\n",
      "Requirement already satisfied: prompt-toolkit<2.0.0,>=1.0.4 in /usr/local/lib/python3.5/dist-packages (from IPython>=3.2.0->scikit-bio==0.5.1)\n",
      "Requirement already satisfied: simplegeneric>0.8 in /usr/local/lib/python3.5/dist-packages (from IPython>=3.2.0->scikit-bio==0.5.1)\n",
      "Requirement already satisfied: pygments in /usr/local/lib/python3.5/dist-packages (from IPython>=3.2.0->scikit-bio==0.5.1)\n",
      "Requirement already satisfied: jedi>=0.10 in /usr/local/lib/python3.5/dist-packages (from IPython>=3.2.0->scikit-bio==0.5.1)\n",
      "Requirement already satisfied: pickleshare in /usr/local/lib/python3.5/dist-packages (from IPython>=3.2.0->scikit-bio==0.5.1)\n",
      "Requirement already satisfied: pexpect; sys_platform != \"win32\" in /usr/local/lib/python3.5/dist-packages (from IPython>=3.2.0->scikit-bio==0.5.1)\n",
      "Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /usr/local/lib/python3.5/dist-packages (from requests->CacheControl>=0.11.5->scikit-bio==0.5.1)\n",
      "Requirement already satisfied: urllib3<1.23,>=1.21.1 in /usr/local/lib/python3.5/dist-packages (from requests->CacheControl>=0.11.5->scikit-bio==0.5.1)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.5/dist-packages (from requests->CacheControl>=0.11.5->scikit-bio==0.5.1)\n",
      "Requirement already satisfied: idna<2.7,>=2.5 in /usr/local/lib/python3.5/dist-packages (from requests->CacheControl>=0.11.5->scikit-bio==0.5.1)\n",
      "Requirement already satisfied: ipython-genutils in /usr/local/lib/python3.5/dist-packages (from traitlets>=4.2->IPython>=3.2.0->scikit-bio==0.5.1)\n",
      "Requirement already satisfied: wcwidth in /usr/local/lib/python3.5/dist-packages (from prompt-toolkit<2.0.0,>=1.0.4->IPython>=3.2.0->scikit-bio==0.5.1)\n",
      "Requirement already satisfied: parso==0.1.0 in /usr/local/lib/python3.5/dist-packages (from jedi>=0.10->IPython>=3.2.0->scikit-bio==0.5.1)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /usr/local/lib/python3.5/dist-packages (from pexpect; sys_platform != \"win32\"->IPython>=3.2.0->scikit-bio==0.5.1)\n"
     ]
    }
   ],
   "source": [
    "!pip install scikit-bio==0.5.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 590,
   "metadata": {
    "scrolled": false,
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [],
   "source": [
    "import skbio.io"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 591,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "seqs = []\n",
    "with open('evasins_uniprot_hits.fasta') as fh:\n",
    "    for seq in skbio.io.read(fh, format='fasta'):\n",
    "        seqs.append(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 592,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "132"
      ]
     },
     "execution_count": 592,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 593,
   "metadata": {
    "scrolled": false,
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence\n",
      "---------------------------------------------------------------------\n",
      "Metadata:\n",
      "    'description': 'Putative evasin 1 OS=Amblyomma triste PE=2 SV=1'\n",
      "    'id': 'tr|A0A023G8N3|A0A023G8N3_9ACAR'\n",
      "Stats:\n",
      "    length: 146\n",
      "---------------------------------------------------------------------\n",
      "0   MTTTTSATLI FLLYVSQLLV GFSRNEASDP PQTDEDCEYY DPAVDNITCT IQSLNTTGDP\n",
      "60  IPVGCLATCE NSTRYLPNGT ECLGISQHVA NRMQGNVNYT CPVGLCYRGV CQRNGLGIDC\n",
      "120 WHDTPPPNST NVTTKAPTTL TSGRDL\n"
     ]
    }
   ],
   "source": [
    "print(repr(seqs[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 594,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "skbio.sequence._sequence.Sequence"
      ]
     },
     "execution_count": 594,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(seqs[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 595,
   "metadata": {
    "scrolled": true,
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on Sequence in module skbio.sequence._sequence object:\n",
      "\n",
      "class Sequence(skbio.metadata._mixin.MetadataMixin, skbio.metadata._mixin.PositionalMetadataMixin, skbio.metadata._mixin.IntervalMetadataMixin, collections.abc.Sequence, skbio._base.SkbioObject)\n",
      " |  Store generic sequence data and optional associated metadata.\n",
      " |  \n",
      " |  ``Sequence`` objects do not enforce an alphabet or grammar and are thus the\n",
      " |  most generic objects for storing sequence data. ``Sequence`` objects do not\n",
      " |  necessarily represent biological sequences. For example, ``Sequence`` can\n",
      " |  be used to represent a position in a multiple sequence alignment.\n",
      " |  Subclasses ``DNA``, ``RNA``, and ``Protein`` enforce the IUPAC character\n",
      " |  set [1]_ for, and provide operations specific to, each respective molecule\n",
      " |  type.\n",
      " |  \n",
      " |  ``Sequence`` objects consist of the underlying sequence data, as well\n",
      " |  as optional metadata and positional metadata. The underlying sequence\n",
      " |  is immutable, while the metdata and positional metadata are mutable.\n",
      " |  \n",
      " |  Parameters\n",
      " |  ----------\n",
      " |  sequence : str, Sequence, or 1D np.ndarray (np.uint8 or '\\|S1')\n",
      " |      Characters representing the sequence itself.\n",
      " |  metadata : dict, optional\n",
      " |      Arbitrary metadata which applies to the entire sequence. A shallow copy\n",
      " |      of the ``dict`` will be made (see Examples section below for details).\n",
      " |  positional_metadata : pd.DataFrame consumable, optional\n",
      " |      Arbitrary per-character metadata (e.g., sequence read quality\n",
      " |      scores). Must be able to be passed directly to ``pd.DataFrame``\n",
      " |      constructor. Each column of metadata must be the same length as\n",
      " |      `sequence`. A shallow copy of the positional metadata will be made if\n",
      " |      necessary (see Examples section below for details).\n",
      " |  interval_metadata : IntervalMetadata\n",
      " |      Arbitrary metadata which applies to intervals within a sequence to\n",
      " |      store interval features (such as genes, ncRNA on the sequence).\n",
      " |  lowercase : bool or str, optional\n",
      " |      If ``True``, lowercase sequence characters will be converted to\n",
      " |      uppercase characters. If ``False``, no characters will be converted.\n",
      " |      If a str, it will be treated as a key into the positional metadata of\n",
      " |      the object. All lowercase characters will be converted to uppercase,\n",
      " |      and a ``True`` value will be stored in a boolean array in the\n",
      " |      positional metadata under the key.\n",
      " |  \n",
      " |  Attributes\n",
      " |  ----------\n",
      " |  values\n",
      " |  metadata\n",
      " |  positional_metadata\n",
      " |  interval_metadata\n",
      " |  observed_chars\n",
      " |  \n",
      " |  See Also\n",
      " |  --------\n",
      " |  DNA\n",
      " |  RNA\n",
      " |  Protein\n",
      " |  \n",
      " |  References\n",
      " |  ----------\n",
      " |  .. [1] Nomenclature for incompletely specified bases in nucleic acid\n",
      " |     sequences: recommendations 1984.\n",
      " |     Nucleic Acids Res. May 10, 1985; 13(9): 3021-3030.\n",
      " |     A Cornish-Bowden\n",
      " |  \n",
      " |  Examples\n",
      " |  --------\n",
      " |  >>> from pprint import pprint\n",
      " |  >>> from skbio import Sequence\n",
      " |  >>> from skbio.metadata import IntervalMetadata\n",
      " |  \n",
      " |  **Creating sequences:**\n",
      " |  \n",
      " |  Create a sequence without any metadata:\n",
      " |  \n",
      " |  >>> seq = Sequence('GGUCGUGAAGGA')\n",
      " |  >>> seq\n",
      " |  Sequence\n",
      " |  ---------------\n",
      " |  Stats:\n",
      " |      length: 12\n",
      " |  ---------------\n",
      " |  0 GGUCGUGAAG GA\n",
      " |  \n",
      " |  Create a sequence with metadata and positional metadata:\n",
      " |  \n",
      " |  >>> metadata = {'id':'seq-id', 'desc':'seq desc', 'authors': ['Alice']}\n",
      " |  >>> positional_metadata = {'quality': [3, 3, 4, 10],\n",
      " |  ...                        'exons': [True, True, False, True]}\n",
      " |  >>> interval_metadata = IntervalMetadata(4)\n",
      " |  >>> interval = interval_metadata.add([(1, 3)], metadata={'gene': 'sagA'})\n",
      " |  >>> seq = Sequence('ACGT', metadata=metadata,\n",
      " |  ...                positional_metadata=positional_metadata,\n",
      " |  ...                interval_metadata=interval_metadata)\n",
      " |  >>> seq\n",
      " |  Sequence\n",
      " |  -----------------------------\n",
      " |  Metadata:\n",
      " |      'authors': <class 'list'>\n",
      " |      'desc': 'seq desc'\n",
      " |      'id': 'seq-id'\n",
      " |  Positional metadata:\n",
      " |      'exons': <dtype: bool>\n",
      " |      'quality': <dtype: int64>\n",
      " |  Interval metadata:\n",
      " |      1 interval feature\n",
      " |  Stats:\n",
      " |      length: 4\n",
      " |  -----------------------------\n",
      " |  0 ACGT\n",
      " |  \n",
      " |  **Retrieving underlying sequence data:**\n",
      " |  \n",
      " |  Retrieve underlying sequence:\n",
      " |  \n",
      " |  >>> seq.values # doctest: +NORMALIZE_WHITESPACE\n",
      " |  array([b'A', b'C', b'G', b'T'],\n",
      " |        dtype='|S1')\n",
      " |  \n",
      " |  Underlying sequence immutable:\n",
      " |  \n",
      " |  >>> seq.values = np.array([b'T', b'C', b'G', b'A'], dtype='|S1')\n",
      " |  Traceback (most recent call last):\n",
      " |      ...\n",
      " |  AttributeError: can't set attribute\n",
      " |  \n",
      " |  >>> seq.values[0] = b'T'\n",
      " |  Traceback (most recent call last):\n",
      " |      ...\n",
      " |  ValueError: assignment destination is read-only\n",
      " |  \n",
      " |  **Retrieving sequence metadata:**\n",
      " |  \n",
      " |  Retrieve metadata:\n",
      " |  \n",
      " |  >>> pprint(seq.metadata) # using pprint to display dict in sorted order\n",
      " |  {'authors': ['Alice'], 'desc': 'seq desc', 'id': 'seq-id'}\n",
      " |  \n",
      " |  Retrieve positional metadata:\n",
      " |  \n",
      " |  >>> seq.positional_metadata\n",
      " |     exons  quality\n",
      " |  0   True        3\n",
      " |  1   True        3\n",
      " |  2  False        4\n",
      " |  3   True       10\n",
      " |  \n",
      " |  Retrieve interval metadata:\n",
      " |  \n",
      " |  >>> seq.interval_metadata   # doctest: +ELLIPSIS\n",
      " |  1 interval feature\n",
      " |  ------------------\n",
      " |  Interval(interval_metadata=<...>, bounds=[(1, 3)], fuzzy=[(False, False)], metadata={'gene': 'sagA'})\n",
      " |  \n",
      " |  **Updating sequence metadata:**\n",
      " |  \n",
      " |  .. warning:: Be aware that a shallow copy of ``metadata`` and\n",
      " |     ``positional_metadata`` is made for performance. Since a deep copy is\n",
      " |     not made, changes made to mutable Python objects stored as metadata may\n",
      " |     affect the metadata of other ``Sequence`` objects or anything else that\n",
      " |     shares a reference to the object. The following examples illustrate this\n",
      " |     behavior.\n",
      " |  \n",
      " |  First, let's create a sequence and update its metadata:\n",
      " |  \n",
      " |  >>> metadata = {'id':'seq-id', 'desc':'seq desc', 'authors': ['Alice']}\n",
      " |  >>> seq = Sequence('ACGT', metadata=metadata)\n",
      " |  >>> seq.metadata['id'] = 'new-id'\n",
      " |  >>> seq.metadata['pubmed'] = 12345\n",
      " |  >>> pprint(seq.metadata)\n",
      " |  {'authors': ['Alice'], 'desc': 'seq desc', 'id': 'new-id', 'pubmed': 12345}\n",
      " |  \n",
      " |  Note that the original metadata dictionary (stored in variable\n",
      " |  ``metadata``) hasn't changed because a shallow copy was made:\n",
      " |  \n",
      " |  >>> pprint(metadata)\n",
      " |  {'authors': ['Alice'], 'desc': 'seq desc', 'id': 'seq-id'}\n",
      " |  >>> seq.metadata == metadata\n",
      " |  False\n",
      " |  \n",
      " |  Note however that since only a *shallow* copy was made, updates to mutable\n",
      " |  objects will also change the original metadata dictionary:\n",
      " |  \n",
      " |  >>> seq.metadata['authors'].append('Bob')\n",
      " |  >>> seq.metadata['authors']\n",
      " |  ['Alice', 'Bob']\n",
      " |  >>> metadata['authors']\n",
      " |  ['Alice', 'Bob']\n",
      " |  \n",
      " |  This behavior can also occur when manipulating a sequence that has been\n",
      " |  derived from another sequence:\n",
      " |  \n",
      " |  >>> subseq = seq[1:3]\n",
      " |  >>> subseq\n",
      " |  Sequence\n",
      " |  -----------------------------\n",
      " |  Metadata:\n",
      " |      'authors': <class 'list'>\n",
      " |      'desc': 'seq desc'\n",
      " |      'id': 'new-id'\n",
      " |      'pubmed': 12345\n",
      " |  Stats:\n",
      " |      length: 2\n",
      " |  -----------------------------\n",
      " |  0 CG\n",
      " |  >>> pprint(subseq.metadata)\n",
      " |  {'authors': ['Alice', 'Bob'],\n",
      " |   'desc': 'seq desc',\n",
      " |   'id': 'new-id',\n",
      " |   'pubmed': 12345}\n",
      " |  \n",
      " |  The subsequence has inherited the metadata of its parent sequence. If we\n",
      " |  update the subsequence's author list, we see the changes propagated in the\n",
      " |  parent sequence and original metadata dictionary:\n",
      " |  \n",
      " |  >>> subseq.metadata['authors'].append('Carol')\n",
      " |  >>> subseq.metadata['authors']\n",
      " |  ['Alice', 'Bob', 'Carol']\n",
      " |  >>> seq.metadata['authors']\n",
      " |  ['Alice', 'Bob', 'Carol']\n",
      " |  >>> metadata['authors']\n",
      " |  ['Alice', 'Bob', 'Carol']\n",
      " |  \n",
      " |  The behavior for updating positional metadata is similar. Let's create a\n",
      " |  new sequence with positional metadata that is already stored in a\n",
      " |  ``pd.DataFrame``:\n",
      " |  \n",
      " |  >>> positional_metadata = pd.DataFrame(\n",
      " |  ...     {'quality': [3, 3, 4, 10], 'list': [[], [], [], []]})\n",
      " |  >>> seq = Sequence('ACGT', positional_metadata=positional_metadata)\n",
      " |  >>> seq\n",
      " |  Sequence\n",
      " |  -----------------------------\n",
      " |  Positional metadata:\n",
      " |      'list': <dtype: object>\n",
      " |      'quality': <dtype: int64>\n",
      " |  Stats:\n",
      " |      length: 4\n",
      " |  -----------------------------\n",
      " |  0 ACGT\n",
      " |  >>> seq.positional_metadata\n",
      " |    list  quality\n",
      " |  0   []        3\n",
      " |  1   []        3\n",
      " |  2   []        4\n",
      " |  3   []       10\n",
      " |  \n",
      " |  Now let's update the sequence's positional metadata by adding a new column\n",
      " |  and changing a value in another column:\n",
      " |  \n",
      " |  >>> seq.positional_metadata['gaps'] = [False, False, False, False]\n",
      " |  >>> seq.positional_metadata.loc[0, 'quality'] = 999\n",
      " |  >>> seq.positional_metadata\n",
      " |    list  quality   gaps\n",
      " |  0   []      999  False\n",
      " |  1   []        3  False\n",
      " |  2   []        4  False\n",
      " |  3   []       10  False\n",
      " |  \n",
      " |  Note that the original positional metadata (stored in variable\n",
      " |  ``positional_metadata``) hasn't changed because a shallow copy was made:\n",
      " |  \n",
      " |  >>> positional_metadata\n",
      " |    list  quality\n",
      " |  0   []        3\n",
      " |  1   []        3\n",
      " |  2   []        4\n",
      " |  3   []       10\n",
      " |  >>> seq.positional_metadata.equals(positional_metadata)\n",
      " |  False\n",
      " |  \n",
      " |  Next let's create a sequence that has been derived from another sequence:\n",
      " |  \n",
      " |  >>> subseq = seq[1:3]\n",
      " |  >>> subseq\n",
      " |  Sequence\n",
      " |  -----------------------------\n",
      " |  Positional metadata:\n",
      " |      'list': <dtype: object>\n",
      " |      'quality': <dtype: int64>\n",
      " |      'gaps': <dtype: bool>\n",
      " |  Stats:\n",
      " |      length: 2\n",
      " |  -----------------------------\n",
      " |  0 CG\n",
      " |  >>> subseq.positional_metadata\n",
      " |    list  quality   gaps\n",
      " |  0   []        3  False\n",
      " |  1   []        4  False\n",
      " |  \n",
      " |  As described above for metadata, since only a *shallow* copy was made of\n",
      " |  the positional metadata, updates to mutable objects will also change the\n",
      " |  parent sequence's positional metadata and the original positional metadata\n",
      " |  ``pd.DataFrame``:\n",
      " |  \n",
      " |  >>> subseq.positional_metadata.loc[0, 'list'].append('item')\n",
      " |  >>> subseq.positional_metadata\n",
      " |       list  quality   gaps\n",
      " |  0  [item]        3  False\n",
      " |  1      []        4  False\n",
      " |  >>> seq.positional_metadata\n",
      " |       list  quality   gaps\n",
      " |  0      []      999  False\n",
      " |  1  [item]        3  False\n",
      " |  2      []        4  False\n",
      " |  3      []       10  False\n",
      " |  >>> positional_metadata\n",
      " |       list  quality\n",
      " |  0      []        3\n",
      " |  1  [item]        3\n",
      " |  2      []        4\n",
      " |  3      []       10\n",
      " |  \n",
      " |  You can also update the interval metadata. Let's re-create a\n",
      " |  ``Sequence`` object with interval metadata at first:\n",
      " |  \n",
      " |  >>> seq = Sequence('ACGT')\n",
      " |  >>> interval = seq.interval_metadata.add(\n",
      " |  ...     [(1, 3)], metadata={'gene': 'foo'})\n",
      " |  \n",
      " |  You can update directly on the ``Interval`` object:\n",
      " |  \n",
      " |  >>> interval  # doctest: +ELLIPSIS\n",
      " |  Interval(interval_metadata=<...>, bounds=[(1, 3)], fuzzy=[(False, False)], metadata={'gene': 'foo'})\n",
      " |  >>> interval.bounds = [(0, 2)]\n",
      " |  >>> interval  # doctest: +ELLIPSIS\n",
      " |  Interval(interval_metadata=<...>, bounds=[(0, 2)], fuzzy=[(False, False)], metadata={'gene': 'foo'})\n",
      " |  \n",
      " |  You can also query and obtain the interval features you are\n",
      " |  interested and then modify them:\n",
      " |  \n",
      " |  >>> intervals = list(seq.interval_metadata.query(metadata={'gene': 'foo'}))\n",
      " |  >>> intervals[0].fuzzy = [(True, False)]\n",
      " |  >>> print(intervals[0])  # doctest: +ELLIPSIS\n",
      " |  Interval(interval_metadata=<...>, bounds=[(0, 2)], fuzzy=[(True, False)], metadata={'gene': 'foo'})\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      Sequence\n",
      " |      skbio.metadata._mixin.MetadataMixin\n",
      " |      skbio.metadata._mixin.PositionalMetadataMixin\n",
      " |      skbio.metadata._mixin.IntervalMetadataMixin\n",
      " |      collections.abc.Sequence\n",
      " |      collections.abc.Sized\n",
      " |      collections.abc.Iterable\n",
      " |      collections.abc.Container\n",
      " |      skbio._base.SkbioObject\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __bool__(self)\n",
      " |      Returns truth value (truthiness) of sequence.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      bool\n",
      " |          True if length of sequence is greater than 0, else False.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> bool(Sequence(''))\n",
      " |      False\n",
      " |      >>> bool(Sequence('ACGT'))\n",
      " |      True\n",
      " |  \n",
      " |  __contains__(self, subsequence)\n",
      " |      Determine if a subsequence is contained in this sequence.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      subsequence : str, Sequence, or 1D np.ndarray (np.uint8 or '\\|S1')\n",
      " |          The putative subsequence.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      bool\n",
      " |          Indicates whether `subsequence` is contained in this sequence.\n",
      " |      \n",
      " |      Raises\n",
      " |      ------\n",
      " |      TypeError\n",
      " |          If `subsequence` is a ``Sequence`` object with a different type\n",
      " |          than this sequence.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> s = Sequence('GGUCGUGAAGGA')\n",
      " |      >>> 'GGU' in s\n",
      " |      True\n",
      " |      >>> 'CCC' in s\n",
      " |      False\n",
      " |  \n",
      " |  __copy__(self)\n",
      " |      Return a shallow copy of this sequence.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      copy\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      This method is equivalent to ``seq.copy(deep=False)``.\n",
      " |  \n",
      " |  __deepcopy__(self, memo)\n",
      " |      Return a deep copy of this sequence.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      copy\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      This method is equivalent to ``seq.copy(deep=True)``.\n",
      " |  \n",
      " |  __eq__(self, other)\n",
      " |      Determine if this sequence is equal to another.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      Sequences are equal if they are *exactly* the same type and their\n",
      " |      sequence characters, metadata, and positional metadata are the same.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      other : Sequence\n",
      " |          Sequence to test for equality against.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      bool\n",
      " |          Indicates whether this sequence is equal to `other`.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      Define two ``Sequence`` objects that have the same underlying sequence\n",
      " |      of characters:\n",
      " |      \n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> s = Sequence('ACGT')\n",
      " |      >>> t = Sequence('ACGT')\n",
      " |      \n",
      " |      The two sequences are considered equal because they are the same type,\n",
      " |      their underlying sequence of characters are the same, and their\n",
      " |      optional metadata attributes (``metadata`` and ``positional_metadata``)\n",
      " |      were not provided:\n",
      " |      \n",
      " |      >>> s == t\n",
      " |      True\n",
      " |      >>> t == s\n",
      " |      True\n",
      " |      \n",
      " |      Define another sequence object with a different sequence of characters\n",
      " |      than the previous two sequence objects:\n",
      " |      \n",
      " |      >>> u = Sequence('ACGA')\n",
      " |      >>> u == t\n",
      " |      False\n",
      " |      \n",
      " |      Define a sequence with the same sequence of characters as ``u`` but\n",
      " |      with different metadata, positional metadata, and interval metadata:\n",
      " |      \n",
      " |      >>> v = Sequence('ACGA', metadata={'id': 'abc'},\n",
      " |      ...              positional_metadata={'quality':[1, 5, 3, 3]})\n",
      " |      >>> _ = v.interval_metadata.add([(0, 1)])\n",
      " |      \n",
      " |      The two sequences are not considered equal because their metadata,\n",
      " |      positional metadata, and interval metadata do not match:\n",
      " |      \n",
      " |      >>> u == v\n",
      " |      False\n",
      " |  \n",
      " |  __getitem__(self, indexable)\n",
      " |      Slice this sequence.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      This drops the ``self.interval_metadata`` from the returned\n",
      " |      new ``Sequence`` object.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      indexable : int, slice, iterable (int and slice), 1D array_like (bool)\n",
      " |          The position(s) to return from this sequence. If `indexable` is an\n",
      " |          iterable of integers, these are assumed to be indices in the\n",
      " |          sequence to keep. If `indexable` is a 1D ``array_like`` of\n",
      " |          booleans, these are assumed to be the positions in the sequence to\n",
      " |          keep.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Sequence\n",
      " |          New sequence containing the position(s) specified by `indexable` in\n",
      " |          this sequence. Positional metadata will be sliced in the same\n",
      " |          manner and included in the returned sequence. `metadata` is\n",
      " |          included in the returned sequence.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> s = Sequence('GGUCGUGAAGGA')\n",
      " |      \n",
      " |      Obtain a single character from the sequence:\n",
      " |      \n",
      " |      >>> s[1]\n",
      " |      Sequence\n",
      " |      -------------\n",
      " |      Stats:\n",
      " |          length: 1\n",
      " |      -------------\n",
      " |      0 G\n",
      " |      \n",
      " |      Obtain a slice:\n",
      " |      \n",
      " |      >>> s[7:]\n",
      " |      Sequence\n",
      " |      -------------\n",
      " |      Stats:\n",
      " |          length: 5\n",
      " |      -------------\n",
      " |      0 AAGGA\n",
      " |      \n",
      " |      Obtain characters at the following indices:\n",
      " |      \n",
      " |      >>> s[[3, 4, 7, 0, 3]]\n",
      " |      Sequence\n",
      " |      -------------\n",
      " |      Stats:\n",
      " |          length: 5\n",
      " |      -------------\n",
      " |      0 CGAGC\n",
      " |      \n",
      " |      Obtain characters at positions evaluating to `True`:\n",
      " |      \n",
      " |      >>> s = Sequence('GGUCG')\n",
      " |      >>> index = [True, False, True, 'a' is 'a', False]\n",
      " |      >>> s[index]\n",
      " |      Sequence\n",
      " |      -------------\n",
      " |      Stats:\n",
      " |          length: 3\n",
      " |      -------------\n",
      " |      0 GUC\n",
      " |  \n",
      " |  __init__(self, sequence, metadata=None, positional_metadata=None, interval_metadata=None, lowercase=False)\n",
      " |      State: Stable as of 0.4.0.\n",
      " |  \n",
      " |  __iter__(self)\n",
      " |      Iterate over positions in this sequence.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      Yields\n",
      " |      ------\n",
      " |      Sequence\n",
      " |          Single character subsequence, one for each position in the\n",
      " |          sequence.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> s = Sequence('GGUC')\n",
      " |      >>> for c in s:\n",
      " |      ...     str(c)\n",
      " |      'G'\n",
      " |      'G'\n",
      " |      'U'\n",
      " |      'C'\n",
      " |  \n",
      " |  __len__(self)\n",
      " |      Return the number of characters in this sequence.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      int\n",
      " |          The length of this sequence.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> s = Sequence('GGUC')\n",
      " |      >>> len(s)\n",
      " |      4\n",
      " |  \n",
      " |  __ne__(self, other)\n",
      " |      Determine if this sequence is not equal to another.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      Sequences are not equal if they are not *exactly* the same type, or\n",
      " |      their sequence characters, metadata, or positional metadata differ.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      other : Sequence\n",
      " |          Sequence to test for inequality against.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      bool\n",
      " |          Indicates whether this sequence is not equal to `other`.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> s = Sequence('ACGT')\n",
      " |      >>> t = Sequence('ACGT')\n",
      " |      >>> s != t\n",
      " |      False\n",
      " |      >>> u = Sequence('ACGA')\n",
      " |      >>> u != t\n",
      " |      True\n",
      " |      >>> v = Sequence('ACGA', metadata={'id': 'v'})\n",
      " |      >>> u != v\n",
      " |      True\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return a string representation of this sequence object.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      Representation includes:\n",
      " |      \n",
      " |      * sequence type\n",
      " |      * metadata keys and values: will display key/value if it is an\n",
      " |        understood type, otherwise just the type will be displayed. If it is\n",
      " |        an understood type whose representation is too long, just the type\n",
      " |        will be displayed\n",
      " |      * positional metadata: column names and column dtypes will be displayed\n",
      " |        in the order they appear in the positional metadata ``pd.DataFrame``.\n",
      " |        Column names (i.e., keys) follow the same display rules as metadata\n",
      " |        keys\n",
      " |      * interval metadata: the number of interval features will be displayed.\n",
      " |      * sequence stats (e.g., length)\n",
      " |      * up to five lines of chunked sequence data. Each line of chunked\n",
      " |        sequence data displays the current position in the sequence\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      str\n",
      " |          String representation of this sequence object.\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      Subclasses can override Sequence._repr_stats to provide custom\n",
      " |      statistics.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      Short sequence without metadata:\n",
      " |      \n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> from skbio.metadata._interval import IntervalMetadata\n",
      " |      >>> Sequence('ACGTAATGGATACGTAATGCA')\n",
      " |      Sequence\n",
      " |      -------------------------\n",
      " |      Stats:\n",
      " |          length: 21\n",
      " |      -------------------------\n",
      " |      0 ACGTAATGGA TACGTAATGC A\n",
      " |      \n",
      " |      Longer sequence displays first two lines and last two lines:\n",
      " |      \n",
      " |      >>> Sequence('ACGT' * 100)\n",
      " |      Sequence\n",
      " |      ---------------------------------------------------------------------\n",
      " |      Stats:\n",
      " |          length: 400\n",
      " |      ---------------------------------------------------------------------\n",
      " |      0   ACGTACGTAC GTACGTACGT ACGTACGTAC GTACGTACGT ACGTACGTAC GTACGTACGT\n",
      " |      60  ACGTACGTAC GTACGTACGT ACGTACGTAC GTACGTACGT ACGTACGTAC GTACGTACGT\n",
      " |      ...\n",
      " |      300 ACGTACGTAC GTACGTACGT ACGTACGTAC GTACGTACGT ACGTACGTAC GTACGTACGT\n",
      " |      360 ACGTACGTAC GTACGTACGT ACGTACGTAC GTACGTACGT\n",
      " |      \n",
      " |      Sequence with metadata, positional metadata, and interval metadata:\n",
      " |      \n",
      " |      >>> metadata = {\n",
      " |      ...     'id': 'seq-id',\n",
      " |      ...     'description': 'description of the sequence, wrapping across '\n",
      " |      ...     'lines if it\\'s too long',\n",
      " |      ...     'authors': ['Alice', 'Bob', 'Carol'],\n",
      " |      ...     'year': 2015,\n",
      " |      ...     'published': True\n",
      " |      ... }\n",
      " |      >>> positional_metadata = {\n",
      " |      ...     'quality': [3, 10, 11, 10],\n",
      " |      ...     'exons': [True, True, False, True]\n",
      " |      ... }\n",
      " |      >>> seq = Sequence('ACGT', metadata=metadata,\n",
      " |      ...          positional_metadata=positional_metadata)\n",
      " |      >>> _ = seq.interval_metadata.add([(0, 2)], metadata={'gene': 'sagA'})\n",
      " |      >>> seq\n",
      " |      Sequence\n",
      " |      ----------------------------------------------------------------------\n",
      " |      Metadata:\n",
      " |          'authors': <class 'list'>\n",
      " |          'description': \"description of the sequence, wrapping across lines\n",
      " |                          if it's too long\"\n",
      " |          'id': 'seq-id'\n",
      " |          'published': True\n",
      " |          'year': 2015\n",
      " |      Positional metadata:\n",
      " |          'exons': <dtype: bool>\n",
      " |          'quality': <dtype: int64>\n",
      " |      Interval metadata:\n",
      " |          1 interval feature\n",
      " |      Stats:\n",
      " |          length: 4\n",
      " |      ----------------------------------------------------------------------\n",
      " |      0 ACGT\n",
      " |  \n",
      " |  __reversed__(self)\n",
      " |      Iterate over positions in this sequence in reverse order.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      Yields\n",
      " |      ------\n",
      " |      Sequence\n",
      " |          Single character subsequence, one for each position in the\n",
      " |          sequence.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> s = Sequence('GGUC')\n",
      " |      >>> for c in reversed(s):\n",
      " |      ...     str(c)\n",
      " |      'C'\n",
      " |      'U'\n",
      " |      'G'\n",
      " |      'G'\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Return sequence characters as a string.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      str\n",
      " |          Sequence characters as a string. No metadata or positional\n",
      " |          metadata will be included.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      sequence\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> s = Sequence('GGUCGUAAAGGA', metadata={'id':'hello'})\n",
      " |      >>> str(s)\n",
      " |      'GGUCGUAAAGGA'\n",
      " |  \n",
      " |  count(self, subsequence, start=None, end=None)\n",
      " |      Count occurrences of a subsequence in this sequence.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      subsequence : str, Sequence, or 1D np.ndarray (np.uint8 or '\\|S1')\n",
      " |          Subsequence to count occurrences of.\n",
      " |      start : int, optional\n",
      " |          The position at which to start counting (inclusive).\n",
      " |      end : int, optional\n",
      " |          The position at which to stop counting (exclusive).\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      int\n",
      " |          Number of occurrences of `subsequence` in this sequence.\n",
      " |      \n",
      " |      Raises\n",
      " |      ------\n",
      " |      ValueError\n",
      " |          If `subsequence` is of length 0.\n",
      " |      TypeError\n",
      " |          If `subsequence` is a ``Sequence`` object with a different type\n",
      " |          than this sequence.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> s = Sequence('GGUCG')\n",
      " |      >>> s.count('G')\n",
      " |      3\n",
      " |      >>> s.count('GG')\n",
      " |      1\n",
      " |      >>> s.count('T')\n",
      " |      0\n",
      " |      >>> s.count('G', 2, 5)\n",
      " |      1\n",
      " |  \n",
      " |  distance(self, other, metric=None)\n",
      " |      Compute the distance to another sequence.\n",
      " |      \n",
      " |      State: Experimental as of 0.4.0.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      other : str, Sequence, or 1D np.ndarray (np.uint8 or '\\|S1')\n",
      " |          Sequence to compute the distance to. If `other` is a ``Sequence``\n",
      " |          object, it must be the same type as this sequence. Other input\n",
      " |          types will be converted into a ``Sequence`` object of the same type\n",
      " |          as this sequence.\n",
      " |      metric : function, optional\n",
      " |          Function used to compute the distance between this sequence and\n",
      " |          `other`. If ``None`` (the default), Hamming distance will be used\n",
      " |          (:func:`skbio.sequence.distance.hamming`). `metric` should take two\n",
      " |          ``skbio.Sequence`` objects and return a ``float``. The sequence\n",
      " |          objects passed to `metric` will be the same type as this sequence.\n",
      " |          See :mod:`skbio.sequence.distance` for other predefined metrics\n",
      " |          that can be supplied via `metric`.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      float\n",
      " |          Distance between this sequence and `other` as defined by `metric`.\n",
      " |      \n",
      " |      Raises\n",
      " |      ------\n",
      " |      TypeError\n",
      " |          If `other` is a ``Sequence`` object with a different type than this\n",
      " |          sequence.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      skbio.sequence.distance\n",
      " |      fraction_diff\n",
      " |      fraction_same\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> s = Sequence('GGUC')\n",
      " |      >>> t = Sequence('AGUC')\n",
      " |      \n",
      " |      Compute Hamming distance (the default metric):\n",
      " |      \n",
      " |      >>> s.distance(t)\n",
      " |      0.25\n",
      " |      \n",
      " |      Use a custom metric:\n",
      " |      \n",
      " |      >>> def custom_metric(s1, s2): return 0.42\n",
      " |      >>> s.distance(t, custom_metric)\n",
      " |      0.42\n",
      " |  \n",
      " |  find_with_regex(self, regex, ignore=None)\n",
      " |      Generate slices for patterns matched by a regular expression.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      regex : str or regular expression object\n",
      " |          String to be compiled into a regular expression, or a pre-\n",
      " |          compiled regular expression object (e.g., from calling\n",
      " |          ``re.compile``).\n",
      " |      ignore : 1D array_like (bool) or iterable (slices or ints), optional\n",
      " |          Indicate the positions to ignore when matching.\n",
      " |      \n",
      " |      Yields\n",
      " |      ------\n",
      " |      slice\n",
      " |          Location where the regular expression matched.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> s = Sequence('AATATACCGGTTATAA')\n",
      " |      >>> for match in s.find_with_regex('(TATA+)'):\n",
      " |      ...     match\n",
      " |      ...     str(s[match])\n",
      " |      slice(2, 6, None)\n",
      " |      'TATA'\n",
      " |      slice(11, 16, None)\n",
      " |      'TATAA'\n",
      " |  \n",
      " |  frequencies(self, chars=None, relative=False)\n",
      " |      Compute frequencies of characters in the sequence.\n",
      " |      \n",
      " |      State: Experimental as of 0.4.1.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      chars : str or set of str, optional\n",
      " |          Characters to compute the frequencies of. May be a ``str``\n",
      " |          containing a single character or a ``set`` of single-character\n",
      " |          strings. If ``None``, frequencies will be computed for all\n",
      " |          characters present in the sequence.\n",
      " |      relative : bool, optional\n",
      " |          If ``True``, return the relative frequency of each character\n",
      " |          instead of its count. If `chars` is provided, relative frequencies\n",
      " |          will be computed with respect to the number of characters in the\n",
      " |          sequence, **not** the total count of characters observed in\n",
      " |          `chars`. Thus, the relative frequencies will not necessarily sum to\n",
      " |          1.0 if `chars` is provided.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      dict\n",
      " |          Frequencies of characters in the sequence.\n",
      " |      \n",
      " |      Raises\n",
      " |      ------\n",
      " |      TypeError\n",
      " |          If `chars` is not a ``str`` or ``set`` of ``str``.\n",
      " |      ValueError\n",
      " |          If `chars` is not a single-character ``str`` or a ``set`` of\n",
      " |          single-character strings.\n",
      " |      ValueError\n",
      " |          If `chars` contains characters outside the allowable range of\n",
      " |          characters in a ``Sequence`` object.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      kmer_frequencies\n",
      " |      iter_kmers\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      If the sequence is empty (i.e., length zero), ``relative=True``,\n",
      " |      **and** `chars` is provided, the relative frequency of each specified\n",
      " |      character will be ``np.nan``.\n",
      " |      \n",
      " |      If `chars` is not provided, this method is equivalent to, but faster\n",
      " |      than, ``seq.kmer_frequencies(k=1)``.\n",
      " |      \n",
      " |      If `chars` is not provided, it is equivalent to, but faster than,\n",
      " |      passing ``chars=seq.observed_chars``.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      Compute character frequencies of a sequence:\n",
      " |      \n",
      " |      >>> from pprint import pprint\n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> seq = Sequence('AGAAGACC')\n",
      " |      >>> freqs = seq.frequencies()\n",
      " |      >>> pprint(freqs) # using pprint to display dict in sorted order\n",
      " |      {'A': 4, 'C': 2, 'G': 2}\n",
      " |      \n",
      " |      Compute relative character frequencies:\n",
      " |      \n",
      " |      >>> freqs = seq.frequencies(relative=True)\n",
      " |      >>> pprint(freqs)\n",
      " |      {'A': 0.5, 'C': 0.25, 'G': 0.25}\n",
      " |      \n",
      " |      Compute relative frequencies of characters A, C, and T:\n",
      " |      \n",
      " |      >>> freqs = seq.frequencies(chars={'A', 'C', 'T'}, relative=True)\n",
      " |      >>> pprint(freqs)\n",
      " |      {'A': 0.5, 'C': 0.25, 'T': 0.0}\n",
      " |      \n",
      " |      Note that since character T is not in the sequence we receive a\n",
      " |      relative frequency of 0.0. The relative frequencies of A and C are\n",
      " |      relative to the number of characters in the sequence (8), **not** the\n",
      " |      number of A and C characters (4 + 2 = 6).\n",
      " |  \n",
      " |  index(self, subsequence, start=None, end=None)\n",
      " |      Find position where subsequence first occurs in the sequence.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      subsequence : str, Sequence, or 1D np.ndarray (np.uint8 or '\\|S1')\n",
      " |          Subsequence to search for in this sequence.\n",
      " |      start : int, optional\n",
      " |          The position at which to start searching (inclusive).\n",
      " |      end : int, optional\n",
      " |          The position at which to stop searching (exclusive).\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      int\n",
      " |          Position where `subsequence` first occurs in this sequence.\n",
      " |      \n",
      " |      Raises\n",
      " |      ------\n",
      " |      ValueError\n",
      " |          If `subsequence` is not present in this sequence.\n",
      " |      TypeError\n",
      " |          If `subsequence` is a ``Sequence`` object with a different type\n",
      " |          than this sequence.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> s = Sequence('ACACGACGTT-')\n",
      " |      >>> s.index('ACG')\n",
      " |      2\n",
      " |  \n",
      " |  iter_contiguous(self, included, min_length=1, invert=False)\n",
      " |      Yield contiguous subsequences based on `included`.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      included : 1D array_like (bool) or iterable (slices or ints)\n",
      " |          `included` is transformed into a flat boolean vector where each\n",
      " |          position will either be included or skipped. All contiguous\n",
      " |          included positions will be yielded as a single region.\n",
      " |      min_length : int, optional\n",
      " |          The minimum length of a subsequence for it to be yielded.\n",
      " |          Default is 1.\n",
      " |      invert : bool, optional\n",
      " |          Whether to invert `included` such that it describes what should be\n",
      " |          skipped instead of included. Default is False.\n",
      " |      \n",
      " |      Yields\n",
      " |      ------\n",
      " |      Sequence\n",
      " |          Contiguous subsequence as indicated by `included`.\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      If slices provide adjacent ranges, then they will be considered the\n",
      " |      same contiguous subsequence.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      Here we use `iter_contiguous` to find all of the contiguous ungapped\n",
      " |      sequences using a boolean vector derived from our DNA sequence.\n",
      " |      \n",
      " |      >>> from skbio import DNA\n",
      " |      >>> s = DNA('AAA--TT-CCCC-G-')\n",
      " |      >>> no_gaps = ~s.gaps()\n",
      " |      >>> for ungapped_subsequence in s.iter_contiguous(no_gaps,\n",
      " |      ...                                               min_length=2):\n",
      " |      ...     print(ungapped_subsequence)\n",
      " |      AAA\n",
      " |      TT\n",
      " |      CCCC\n",
      " |      \n",
      " |      Note how the last potential subsequence was skipped because it would\n",
      " |      have been smaller than our `min_length` which was set to 2.\n",
      " |      \n",
      " |      We can also use `iter_contiguous` on a generator of slices as is\n",
      " |      produced by `find_motifs` (and `find_with_regex`).\n",
      " |      \n",
      " |      >>> from skbio import Protein\n",
      " |      >>> s = Protein('ACDFNASANFTACGNPNRTESL')\n",
      " |      >>> for subseq in s.iter_contiguous(s.find_motifs('N-glycosylation')):\n",
      " |      ...     print(subseq)\n",
      " |      NASANFTA\n",
      " |      NRTE\n",
      " |      \n",
      " |      Note how the first subsequence contains two N-glycosylation sites. This\n",
      " |      happened because they were contiguous.\n",
      " |  \n",
      " |  iter_kmers(self, k, overlap=True)\n",
      " |      Generate kmers of length `k` from this sequence.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      k : int\n",
      " |          The kmer length.\n",
      " |      overlap : bool, optional\n",
      " |          Defines whether the kmers should be overlapping or not.\n",
      " |      \n",
      " |      Yields\n",
      " |      ------\n",
      " |      Sequence\n",
      " |          kmer of length `k` contained in this sequence.\n",
      " |      \n",
      " |      Raises\n",
      " |      ------\n",
      " |      ValueError\n",
      " |          If `k` is less than 1.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> s = Sequence('ACACGACGTT')\n",
      " |      >>> for kmer in s.iter_kmers(4, overlap=False):\n",
      " |      ...     str(kmer)\n",
      " |      'ACAC'\n",
      " |      'GACG'\n",
      " |      >>> for kmer in s.iter_kmers(3, overlap=True):\n",
      " |      ...     str(kmer)\n",
      " |      'ACA'\n",
      " |      'CAC'\n",
      " |      'ACG'\n",
      " |      'CGA'\n",
      " |      'GAC'\n",
      " |      'ACG'\n",
      " |      'CGT'\n",
      " |      'GTT'\n",
      " |  \n",
      " |  kmer_frequencies(self, k, overlap=True, relative=False)\n",
      " |      Return counts of words of length `k` from this sequence.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      k : int\n",
      " |          The word length.\n",
      " |      overlap : bool, optional\n",
      " |          Defines whether the kmers should be overlapping or not.\n",
      " |      relative : bool, optional\n",
      " |          If ``True``, return the relative frequency of each kmer instead of\n",
      " |          its count.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      dict\n",
      " |          Frequencies of words of length `k` contained in this sequence.\n",
      " |      \n",
      " |      Raises\n",
      " |      ------\n",
      " |      ValueError\n",
      " |          If `k` is less than 1.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> from pprint import pprint\n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> s = Sequence('ACACATTTATTA')\n",
      " |      >>> freqs = s.kmer_frequencies(3, overlap=False)\n",
      " |      >>> pprint(freqs) # using pprint to display dict in sorted order\n",
      " |      {'ACA': 1, 'CAT': 1, 'TTA': 2}\n",
      " |      >>> freqs = s.kmer_frequencies(3, relative=True, overlap=False)\n",
      " |      >>> pprint(freqs)\n",
      " |      {'ACA': 0.25, 'CAT': 0.25, 'TTA': 0.5}\n",
      " |  \n",
      " |  lowercase(self, lowercase)\n",
      " |      Return a case-sensitive string representation of the sequence.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      lowercase: str or boolean vector\n",
      " |          If lowercase is a boolean vector, it is used to set sequence\n",
      " |          characters to lowercase in the output string. True values in the\n",
      " |          boolean vector correspond to lowercase characters. If lowercase\n",
      " |          is a str, it is treated like a key into the positional metadata,\n",
      " |          pointing to a column which must be a boolean vector.\n",
      " |          That boolean vector is then used as described previously.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      str\n",
      " |          String representation of sequence with specified characters set to\n",
      " |          lowercase.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> s = Sequence('ACGT')\n",
      " |      >>> s.lowercase([True, True, False, False])\n",
      " |      'acGT'\n",
      " |      >>> s = Sequence('ACGT',\n",
      " |      ...              positional_metadata={\n",
      " |      ...                 'exons': [True, False, False, True]})\n",
      " |      >>> s.lowercase('exons')\n",
      " |      'aCGt'\n",
      " |      \n",
      " |      Constructor automatically populates a column in positional metadata\n",
      " |      when the ``lowercase`` keyword argument is provided with a column name:\n",
      " |      \n",
      " |      >>> s = Sequence('ACgt', lowercase='introns')\n",
      " |      >>> s.lowercase('introns')\n",
      " |      'ACgt'\n",
      " |      >>> s = Sequence('ACGT', lowercase='introns')\n",
      " |      >>> s.lowercase('introns')\n",
      " |      'ACGT'\n",
      " |  \n",
      " |  match_frequency(self, other, relative=False)\n",
      " |      Return count of positions that are the same between two sequences.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      other : str, Sequence, or 1D np.ndarray (np.uint8 or '\\|S1')\n",
      " |          Sequence to compare to.\n",
      " |      relative : bool, optional\n",
      " |          If ``True``, return the relative frequency of matches instead of\n",
      " |          the count.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      int or float\n",
      " |          Number of positions that are the same between the sequences. This\n",
      " |          will be an ``int`` if `relative` is ``False`` and a ``float``\n",
      " |          if `relative` is ``True``.\n",
      " |      \n",
      " |      Raises\n",
      " |      ------\n",
      " |      ValueError\n",
      " |          If the sequences are not the same length.\n",
      " |      TypeError\n",
      " |          If `other` is a ``Sequence`` object with a different type than this\n",
      " |          sequence.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      mismatch_frequency\n",
      " |      matches\n",
      " |      mismatches\n",
      " |      distance\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> s = Sequence('GGUC')\n",
      " |      >>> t = Sequence('AGUC')\n",
      " |      >>> s.match_frequency(t)\n",
      " |      3\n",
      " |      >>> s.match_frequency(t, relative=True)\n",
      " |      0.75\n",
      " |  \n",
      " |  matches(self, other)\n",
      " |      Find positions that match with another sequence.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      other : str, Sequence, or 1D np.ndarray (np.uint8 or '\\|S1')\n",
      " |          Sequence to compare to.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      1D np.ndarray (bool)\n",
      " |          Boolean vector where ``True`` at position ``i`` indicates a match\n",
      " |          between the sequences at their positions ``i``.\n",
      " |      \n",
      " |      Raises\n",
      " |      ------\n",
      " |      ValueError\n",
      " |          If the sequences are not the same length.\n",
      " |      TypeError\n",
      " |          If `other` is a ``Sequence`` object with a different type than this\n",
      " |          sequence.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      mismatches\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> s = Sequence('GGUC')\n",
      " |      >>> t = Sequence('GAUU')\n",
      " |      >>> s.matches(t)\n",
      " |      array([ True, False,  True, False], dtype=bool)\n",
      " |  \n",
      " |  mismatch_frequency(self, other, relative=False)\n",
      " |      Return count of positions that differ between two sequences.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      other : str, Sequence, or 1D np.ndarray (np.uint8 or '\\|S1')\n",
      " |          Sequence to compare to.\n",
      " |      relative : bool, optional\n",
      " |          If ``True``, return the relative frequency of mismatches instead of\n",
      " |          the count.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      int or float\n",
      " |          Number of positions that differ between the sequences. This will be\n",
      " |          an ``int`` if `relative` is ``False`` and a ``float``\n",
      " |          if `relative` is ``True``.\n",
      " |      \n",
      " |      Raises\n",
      " |      ------\n",
      " |      ValueError\n",
      " |          If the sequences are not the same length.\n",
      " |      TypeError\n",
      " |          If `other` is a ``Sequence`` object with a different type than this\n",
      " |          sequence.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      match_frequency\n",
      " |      matches\n",
      " |      mismatches\n",
      " |      distance\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> s = Sequence('GGUC')\n",
      " |      >>> t = Sequence('AGUC')\n",
      " |      >>> s.mismatch_frequency(t)\n",
      " |      1\n",
      " |      >>> s.mismatch_frequency(t, relative=True)\n",
      " |      0.25\n",
      " |  \n",
      " |  mismatches(self, other)\n",
      " |      Find positions that do not match with another sequence.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      other : str, Sequence, or 1D np.ndarray (np.uint8 or '\\|S1')\n",
      " |          Sequence to compare to.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      1D np.ndarray (bool)\n",
      " |          Boolean vector where ``True`` at position ``i`` indicates a\n",
      " |          mismatch between the sequences at their positions ``i``.\n",
      " |      \n",
      " |      Raises\n",
      " |      ------\n",
      " |      ValueError\n",
      " |          If the sequences are not the same length.\n",
      " |      TypeError\n",
      " |          If `other` is a ``Sequence`` object with a different type than this\n",
      " |          sequence.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      matches\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> s = Sequence('GGUC')\n",
      " |      >>> t = Sequence('GAUU')\n",
      " |      >>> s.mismatches(t)\n",
      " |      array([False,  True, False,  True], dtype=bool)\n",
      " |  \n",
      " |  replace(self, where, character)\n",
      " |      Replace values in this sequence with a different character.\n",
      " |      \n",
      " |      State: Experimental as of 0.5.0.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      where : 1D array_like (bool) or iterable (slices or ints) or str\n",
      " |          Indicates positions in the sequence to replace with `character`.\n",
      " |          Can be a boolean vector, an iterable of indices/slices, or a\n",
      " |          string that is a key in `positional_metadata` pointing to a\n",
      " |          boolean vector.\n",
      " |      character : str or bytes\n",
      " |          Character that will replace chosen items in this sequence.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Sequence\n",
      " |          Copy of this sequence, with chosen items replaced with chosen\n",
      " |          character. All metadata is retained.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      Let's create and display a Sequence:\n",
      " |      \n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> sequence = Sequence('GGTACCAACG')\n",
      " |      >>> str(sequence)\n",
      " |      'GGTACCAACG'\n",
      " |      \n",
      " |      Let's call ``replace`` on the Sequence using a boolean vector for\n",
      " |      ``where`` and assign it to a new variable:\n",
      " |      \n",
      " |      >>> seq = sequence.replace([False, False, False, True, False, False,\n",
      " |      ...                         True, True, False, False], '-')\n",
      " |      \n",
      " |      Let's take a look at the new Sequence:\n",
      " |      \n",
      " |      >>> str(seq)\n",
      " |      'GGT-CC--CG'\n",
      " |      \n",
      " |      Other types of input are accepted by the ``where`` parameter. Let's\n",
      " |      pass in a list of indices and slices that is equivalent to the boolean\n",
      " |      vector we used previously:\n",
      " |      \n",
      " |      >>> str(seq) == str(sequence.replace([3, slice(6, 8)], '-'))\n",
      " |      True\n",
      " |      \n",
      " |      ``where`` also accepts a boolean vector contained in\n",
      " |      ``Sequence.positional_metadata``:\n",
      " |      \n",
      " |      >>> sequence.positional_metadata = {'where':\n",
      " |      ...                                 [False, False, False, True, False,\n",
      " |      ...                                  False, True, True, False, False]}\n",
      " |      \n",
      " |      Let's pass in the key ``'where'`` and compare to ``seq``:\n",
      " |      \n",
      " |      >>> str(seq) == str(sequence.replace('where', '-'))\n",
      " |      True\n",
      " |  \n",
      " |  write(self, file, format='fasta', **kwargs)\n",
      " |      Write an instance of ``Sequence`` to a file.\n",
      " |      \n",
      " |      This is a convenience method for :func:`skbio.io.registry.write`.\n",
      " |      For more information about the I/O system in scikit-bio, please\n",
      " |      see :mod:`skbio.io`.\n",
      " |      \n",
      " |      Supported file formats include:\n",
      " |      \n",
      " |      - ``'fastq'`` (:mod:`skbio.io.format.fastq`)\n",
      " |      - ``'fasta'`` (:mod:`skbio.io.format.fasta`)\n",
      " |      - ``'genbank'`` (:mod:`skbio.io.format.genbank`)\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      file : openable (filepath, URL, filehandle, etc.)\n",
      " |          The location to write the given `format` into.  Something\n",
      " |          that is understood by :func:`skbio.io.util.open`. Filehandles\n",
      " |          are not automatically closed, it is the responsibility of the\n",
      " |          caller.\n",
      " |      format : str\n",
      " |          The format must be a registered format name with a writer for\n",
      " |          ``Sequence``.\n",
      " |          Default is `'fasta'`.\n",
      " |      kwargs : dict, optional\n",
      " |          Keyword arguments passed to :func:`skbio.io.registry.write`\n",
      " |          and the file format writer.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      read\n",
      " |      skbio.io.registry.write\n",
      " |      skbio.io.util.open\n",
      " |      skbio.io.format.fastq\n",
      " |      skbio.io.format.fasta\n",
      " |      skbio.io.format.genbank\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Class methods defined here:\n",
      " |  \n",
      " |  concat(sequences, how='strict') from abc.ABCMeta\n",
      " |      Concatenate an iterable of ``Sequence`` objects.\n",
      " |      \n",
      " |      State: Experimental as of 0.4.1.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      sequences : iterable (Sequence)\n",
      " |          An iterable of ``Sequence`` objects or appropriate subclasses.\n",
      " |      how : {'strict', 'inner', 'outer'}, optional\n",
      " |          How to intersect the `positional_metadata` of the sequences.\n",
      " |          If 'strict': the `positional_metadata` must have the exact same\n",
      " |          columns; 'inner': an inner-join of the columns (only the shared set\n",
      " |          of columns are used); 'outer': an outer-join of the columns\n",
      " |          (all columns are used: missing values will be padded with NaN).\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Sequence\n",
      " |          The returned sequence will be an instance of the class which\n",
      " |          called this class-method.\n",
      " |      \n",
      " |      Raises\n",
      " |      ------\n",
      " |      ValueError\n",
      " |          If `how` is not one of: 'strict', 'inner', or 'outer'.\n",
      " |      ValueError\n",
      " |          If `how` is 'strict' and the `positional_metadata` of each sequence\n",
      " |          does not have the same columns.\n",
      " |      TypeError\n",
      " |          If the sequences cannot be cast as the calling class.\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      The sequence-wide metadata (``Sequence.metadata``) is not retained\n",
      " |      during concatenation.\n",
      " |      \n",
      " |      Sequence objects can be cast to a different type only when the new\n",
      " |      type is an ancestor or child of the original type. Casting between\n",
      " |      sibling types is not allowed, e.g. ``DNA`` -> ``RNA`` is not\n",
      " |      allowed, but ``DNA`` -> ``Sequence`` or ``Sequence`` -> ``DNA``\n",
      " |      would be.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      Concatenate two DNA sequences into a new DNA object:\n",
      " |      \n",
      " |      >>> from skbio import DNA, Sequence\n",
      " |      >>> s1 = DNA(\"ACGT\")\n",
      " |      >>> s2 = DNA(\"GGAA\")\n",
      " |      >>> DNA.concat([s1, s2])\n",
      " |      DNA\n",
      " |      --------------------------\n",
      " |      Stats:\n",
      " |          length: 8\n",
      " |          has gaps: False\n",
      " |          has degenerates: False\n",
      " |          has definites: True\n",
      " |          GC-content: 50.00%\n",
      " |      --------------------------\n",
      " |      0 ACGTGGAA\n",
      " |      \n",
      " |      Concatenate DNA sequences into a Sequence object (type coercion):\n",
      " |      \n",
      " |      >>> Sequence.concat([s1, s2])\n",
      " |      Sequence\n",
      " |      -------------\n",
      " |      Stats:\n",
      " |          length: 8\n",
      " |      -------------\n",
      " |      0 ACGTGGAA\n",
      " |      \n",
      " |      Positional metadata is conserved:\n",
      " |      \n",
      " |      >>> s1 = DNA('AcgT', lowercase='one')\n",
      " |      >>> s2 = DNA('GGaA', lowercase='one',\n",
      " |      ...          positional_metadata={'two': [1, 2, 3, 4]})\n",
      " |      >>> result = DNA.concat([s1, s2], how='outer')\n",
      " |      >>> result\n",
      " |      DNA\n",
      " |      ---------------------------\n",
      " |      Positional metadata:\n",
      " |          'one': <dtype: bool>\n",
      " |          'two': <dtype: float64>\n",
      " |      Stats:\n",
      " |          length: 8\n",
      " |          has gaps: False\n",
      " |          has degenerates: False\n",
      " |          has definites: True\n",
      " |          GC-content: 50.00%\n",
      " |      ---------------------------\n",
      " |      0 ACGTGGAA\n",
      " |      >>> result.positional_metadata\n",
      " |           one  two\n",
      " |      0  False  NaN\n",
      " |      1   True  NaN\n",
      " |      2   True  NaN\n",
      " |      3  False  NaN\n",
      " |      4  False  1.0\n",
      " |      5  False  2.0\n",
      " |      6   True  3.0\n",
      " |      7  False  4.0\n",
      " |  \n",
      " |  read(file, format=None, **kwargs) from abc.ABCMeta\n",
      " |      Create a new ``Sequence`` instance from a file.\n",
      " |      \n",
      " |      This is a convenience method for :func:`skbio.io.registry.read`. For\n",
      " |      more information about the I/O system in scikit-bio, please see\n",
      " |      :mod:`skbio.io`.\n",
      " |      \n",
      " |      Supported file formats include:\n",
      " |      \n",
      " |      - ``'fastq'`` (:mod:`skbio.io.format.fastq`)\n",
      " |      - ``'qseq'`` (:mod:`skbio.io.format.qseq`)\n",
      " |      - ``'fasta'`` (:mod:`skbio.io.format.fasta`)\n",
      " |      - ``'genbank'`` (:mod:`skbio.io.format.genbank`)\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      file : openable (filepath, URL, filehandle, etc.)\n",
      " |          The location to read the given `format`. Something that is\n",
      " |          understood by :func:`skbio.io.util.open`. Filehandles are not\n",
      " |          automatically closed, it is the responsibility of the caller.\n",
      " |      format : str, optional\n",
      " |          The format must be a format name with a reader for ``Sequence``.\n",
      " |          If a `format` is not provided or is None, it will attempt to\n",
      " |          guess the format.\n",
      " |      kwargs : dict, optional\n",
      " |          Keyword arguments passed to :func:`skbio.io.registry.read` and\n",
      " |          the file format reader for ``Sequence``.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Sequence\n",
      " |          A new instance.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      write\n",
      " |      skbio.io.registry.read\n",
      " |      skbio.io.util.open\n",
      " |      skbio.io.format.fastq\n",
      " |      skbio.io.format.qseq\n",
      " |      skbio.io.format.fasta\n",
      " |      skbio.io.format.genbank\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __array_interface__\n",
      " |      Array interface for compatibility with numpy.\n",
      " |      \n",
      " |      This property allows a ``Sequence`` object to share its underlying data\n",
      " |      buffer (``Sequence.values``) with numpy. See [1]_ for more details.\n",
      " |      \n",
      " |      References\n",
      " |      ----------\n",
      " |      .. [1] http://docs.scipy.org/doc/numpy/reference/arrays.interface.html\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> import numpy as np\n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> seq = Sequence('ABC123')\n",
      " |      >>> np.asarray(seq) # doctest: +NORMALIZE_WHITESPACE\n",
      " |      array([b'A', b'B', b'C', b'1', b'2', b'3'],\n",
      " |            dtype='|S1')\n",
      " |  \n",
      " |  observed_chars\n",
      " |      Set of observed characters in the sequence.\n",
      " |      \n",
      " |      State: Experimental as of 0.4.1.\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      This property is not writeable.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> s = Sequence('AACGAC')\n",
      " |      >>> s.observed_chars == {'G', 'A', 'C'}\n",
      " |      True\n",
      " |  \n",
      " |  values\n",
      " |      Array containing underlying sequence characters.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      This property is not writeable.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> s = Sequence('AACGA')\n",
      " |      >>> s.values # doctest: +NORMALIZE_WHITESPACE\n",
      " |      array([b'A', b'A', b'C', b'G', b'A'],\n",
      " |            dtype='|S1')\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  __abstractmethods__ = frozenset()\n",
      " |  \n",
      " |  __hash__ = None\n",
      " |  \n",
      " |  default_write_format = 'fasta'\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from skbio.metadata._mixin.MetadataMixin:\n",
      " |  \n",
      " |  has_metadata(self)\n",
      " |      Determine if the object has metadata.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      An object has metadata if its ``metadata`` dictionary is not empty\n",
      " |      (i.e., has at least one key-value pair).\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      bool\n",
      " |          Indicates whether the object has metadata.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      .. note:: scikit-bio objects with metadata share a common interface for\n",
      " |         accessing and manipulating their metadata. The following examples\n",
      " |         use scikit-bio's ``Sequence`` class to demonstrate metadata\n",
      " |         behavior. These examples apply to all other scikit-bio objects\n",
      " |         storing metadata.\n",
      " |      \n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> seq = Sequence('ACGT')\n",
      " |      >>> seq.has_metadata()\n",
      " |      False\n",
      " |      >>> seq = Sequence('ACGT', metadata={})\n",
      " |      >>> seq.has_metadata()\n",
      " |      False\n",
      " |      >>> seq = Sequence('ACGT', metadata={'id': 'seq-id'})\n",
      " |      >>> seq.has_metadata()\n",
      " |      True\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from skbio.metadata._mixin.MetadataMixin:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  metadata\n",
      " |      ``dict`` containing metadata which applies to the entire object.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      This property can be set and deleted. When setting new metadata a\n",
      " |      shallow copy of the dictionary is made.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      .. note:: scikit-bio objects with metadata share a common interface for\n",
      " |         accessing and manipulating their metadata. The following examples\n",
      " |         use scikit-bio's ``Sequence`` class to demonstrate metadata\n",
      " |         behavior. These examples apply to all other scikit-bio objects\n",
      " |         storing metadata.\n",
      " |      \n",
      " |      Create a sequence with metadata:\n",
      " |      \n",
      " |      >>> from pprint import pprint\n",
      " |      >>> from skbio import Sequence\n",
      " |      >>> seq = Sequence('ACGT', metadata={'id': 'seq-id',\n",
      " |      ...                                  'description': 'seq description'})\n",
      " |      \n",
      " |      Retrieve metadata:\n",
      " |      \n",
      " |      >>> pprint(seq.metadata) # using pprint to display dict in sorted order\n",
      " |      {'description': 'seq description', 'id': 'seq-id'}\n",
      " |      \n",
      " |      Update metadata:\n",
      " |      \n",
      " |      >>> seq.metadata['id'] = 'new-id'\n",
      " |      >>> seq.metadata['pubmed'] = 12345\n",
      " |      >>> pprint(seq.metadata)\n",
      " |      {'description': 'seq description', 'id': 'new-id', 'pubmed': 12345}\n",
      " |      \n",
      " |      Set metadata:\n",
      " |      \n",
      " |      >>> seq.metadata = {'abc': 123}\n",
      " |      >>> seq.metadata\n",
      " |      {'abc': 123}\n",
      " |      \n",
      " |      Delete metadata:\n",
      " |      \n",
      " |      >>> seq.has_metadata()\n",
      " |      True\n",
      " |      >>> del seq.metadata\n",
      " |      >>> seq.metadata\n",
      " |      {}\n",
      " |      >>> seq.has_metadata()\n",
      " |      False\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from skbio.metadata._mixin.PositionalMetadataMixin:\n",
      " |  \n",
      " |  has_positional_metadata(self)\n",
      " |      Determine if the object has positional metadata.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      An object has positional metadata if its ``positional_metadata``\n",
      " |      ``pd.DataFrame`` has at least one column.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      bool\n",
      " |          Indicates whether the object has positional metadata.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      .. note:: scikit-bio objects with positional metadata share a common\n",
      " |         interface for accessing and manipulating their positional metadata.\n",
      " |         The following examples use scikit-bio's ``DNA`` class to demonstrate\n",
      " |         positional metadata behavior. These examples apply to all other\n",
      " |         scikit-bio objects storing positional metadata.\n",
      " |      \n",
      " |      >>> import pandas as pd\n",
      " |      >>> from skbio import DNA\n",
      " |      >>> seq = DNA('ACGT')\n",
      " |      >>> seq.has_positional_metadata()\n",
      " |      False\n",
      " |      >>> seq = DNA('ACGT', positional_metadata=pd.DataFrame(index=range(4)))\n",
      " |      >>> seq.has_positional_metadata()\n",
      " |      False\n",
      " |      >>> seq = DNA('ACGT', positional_metadata={'quality': range(4)})\n",
      " |      >>> seq.has_positional_metadata()\n",
      " |      True\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from skbio.metadata._mixin.PositionalMetadataMixin:\n",
      " |  \n",
      " |  positional_metadata\n",
      " |      ``pd.DataFrame`` containing metadata along an axis.\n",
      " |      \n",
      " |      State: Stable as of 0.4.0.\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      This property can be set and deleted. When setting new positional\n",
      " |      metadata, a shallow copy is made and the ``pd.DataFrame`` index is set\n",
      " |      to ``pd.RangeIndex(start=0, stop=axis_len, step=1)``.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      .. note:: scikit-bio objects with positional metadata share a common\n",
      " |         interface for accessing and manipulating their positional metadata.\n",
      " |         The following examples use scikit-bio's ``DNA`` class to demonstrate\n",
      " |         positional metadata behavior. These examples apply to all other\n",
      " |         scikit-bio objects storing positional metadata.\n",
      " |      \n",
      " |      Create a DNA sequence with positional metadata:\n",
      " |      \n",
      " |      >>> from skbio import DNA\n",
      " |      >>> seq = DNA(\n",
      " |      ...     'ACGT',\n",
      " |      ...     positional_metadata={'quality': [3, 3, 20, 11],\n",
      " |      ...                          'exons': [True, True, False, True]})\n",
      " |      >>> seq\n",
      " |      DNA\n",
      " |      -----------------------------\n",
      " |      Positional metadata:\n",
      " |          'exons': <dtype: bool>\n",
      " |          'quality': <dtype: int64>\n",
      " |      Stats:\n",
      " |          length: 4\n",
      " |          has gaps: False\n",
      " |          has degenerates: False\n",
      " |          has definites: True\n",
      " |          GC-content: 50.00%\n",
      " |      -----------------------------\n",
      " |      0 ACGT\n",
      " |      \n",
      " |      Retrieve positional metadata:\n",
      " |      \n",
      " |      >>> seq.positional_metadata\n",
      " |         exons  quality\n",
      " |      0   True        3\n",
      " |      1   True        3\n",
      " |      2  False       20\n",
      " |      3   True       11\n",
      " |      \n",
      " |      Update positional metadata:\n",
      " |      \n",
      " |      >>> seq.positional_metadata['gaps'] = seq.gaps()\n",
      " |      >>> seq.positional_metadata\n",
      " |         exons  quality   gaps\n",
      " |      0   True        3  False\n",
      " |      1   True        3  False\n",
      " |      2  False       20  False\n",
      " |      3   True       11  False\n",
      " |      \n",
      " |      Set positional metadata:\n",
      " |      \n",
      " |      >>> seq.positional_metadata = {'degenerates': seq.degenerates()}\n",
      " |      >>> seq.positional_metadata\n",
      " |        degenerates\n",
      " |      0       False\n",
      " |      1       False\n",
      " |      2       False\n",
      " |      3       False\n",
      " |      \n",
      " |      Delete positional metadata:\n",
      " |      \n",
      " |      >>> seq.has_positional_metadata()\n",
      " |      True\n",
      " |      >>> del seq.positional_metadata\n",
      " |      >>> seq.positional_metadata\n",
      " |      Empty DataFrame\n",
      " |      Columns: []\n",
      " |      Index: [0, 1, 2, 3]\n",
      " |      >>> seq.has_positional_metadata()\n",
      " |      False\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from skbio.metadata._mixin.IntervalMetadataMixin:\n",
      " |  \n",
      " |  has_interval_metadata(self)\n",
      " |      Determine if the object has interval metadata.\n",
      " |      \n",
      " |      State: Experimental as of 0.5.1.\n",
      " |      \n",
      " |      An object has interval metadata if its ``interval_metadata``\n",
      " |      has at least one ```Interval`` objects.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      bool\n",
      " |          Indicates whether the object has interval metadata.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from skbio.metadata._mixin.IntervalMetadataMixin:\n",
      " |  \n",
      " |  interval_metadata\n",
      " |      ``IntervalMetadata`` object containing info about interval features.\n",
      " |      \n",
      " |      State: Experimental as of 0.5.1.\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      This property can be set and deleted. When setting new\n",
      " |      interval metadata, a shallow copy of the ``IntervalMetadata``\n",
      " |      object is made.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Class methods inherited from collections.abc.Sized:\n",
      " |  \n",
      " |  __subclasshook__(C) from abc.ABCMeta\n",
      " |      Abstract classes can override this to customize issubclass().\n",
      " |      \n",
      " |      This is invoked early on by abc.ABCMeta.__subclasscheck__().\n",
      " |      It should return True, False or NotImplemented.  If it returns\n",
      " |      NotImplemented, the normal algorithm is used.  Otherwise, it\n",
      " |      overrides the normal algorithm (and the outcome is cached).\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(seqs[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Or, look at the nicely formatted version online: http://scikit-bio.org/docs/latest/generated/skbio.sequence.Sequence.html#skbio.sequence.Sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 596,
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'description': 'Putative evasin 1 OS=Amblyomma triste PE=2 SV=1',\n",
       " 'id': 'tr|A0A023G8N3|A0A023G8N3_9ACAR'}"
      ]
     },
     "execution_count": 596,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqs[0].metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 597,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([b'M', b'T', b'T', b'T', b'T', b'S', b'A', b'T', b'L', b'I', b'F',\n",
       "       b'L', b'L', b'Y', b'V', b'S', b'Q', b'L', b'L', b'V', b'G', b'F',\n",
       "       b'S', b'R', b'N', b'E', b'A', b'S', b'D', b'P', b'P', b'Q', b'T',\n",
       "       b'D', b'E', b'D', b'C', b'E', b'Y', b'Y', b'D', b'P', b'A', b'V',\n",
       "       b'D', b'N', b'I', b'T', b'C', b'T', b'I', b'Q', b'S', b'L', b'N',\n",
       "       b'T', b'T', b'G', b'D', b'P', b'I', b'P', b'V', b'G', b'C', b'L',\n",
       "       b'A', b'T', b'C', b'E', b'N', b'S', b'T', b'R', b'Y', b'L', b'P',\n",
       "       b'N', b'G', b'T', b'E', b'C', b'L', b'G', b'I', b'S', b'Q', b'H',\n",
       "       b'V', b'A', b'N', b'R', b'M', b'Q', b'G', b'N', b'V', b'N', b'Y',\n",
       "       b'T', b'C', b'P', b'V', b'G', b'L', b'C', b'Y', b'R', b'G', b'V',\n",
       "       b'C', b'Q', b'R', b'N', b'G', b'L', b'G', b'I', b'D', b'C', b'W',\n",
       "       b'H', b'D', b'T', b'P', b'P', b'P', b'N', b'S', b'T', b'N', b'V',\n",
       "       b'T', b'T', b'K', b'A', b'P', b'T', b'T', b'L', b'T', b'S', b'G',\n",
       "       b'R', b'D', b'L'],\n",
       "      dtype='|S1')"
      ]
     },
     "execution_count": 597,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# A numpy array of each character in the sequence\n",
    "seqs[0].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 598,
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MTTTTSATLIFLLYVSQLLVGFSRNEASDPPQTDEDCEYYDPAVDNITCTIQSLNTTGDPIPVGCLATCENSTRYLPNGTECLGISQHVANRMQGNVNYTCPVGLCYRGVCQRNGLGIDCWHDTPPPNSTNVTTKAPTTLTSGRDL'"
      ]
     },
     "execution_count": 598,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sequence implements a magic __str__ method that allows it to be converted to a string representation\n",
    "str(seqs[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "It's not uncommon for FASTA sequences of protein to include 'X' residues when a residue type is undetermined.\n",
    "\n",
    "Let's see if there are any 'X' residues in sequences of our set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 600,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tr|C9W1Q7|C9W1Q7_RHISA, tr|L7LT68|L7LT68_9ACAR\n",
      "Total sequences: 132\n",
      "Sequences with X residues 2\n"
     ]
    }
   ],
   "source": [
    "x_seqs = []\n",
    "for seq in seqs:\n",
    "    if seq.count('X') >= 1:\n",
    "        x_seqs.append(seq)\n",
    "        \n",
    "print(', '.join([s.metadata['id'] for s in x_seqs]))\n",
    "\n",
    "print(\"Total sequences:\", len(seqs))\n",
    "print(\"Sequences with X residues\", len(x_seqs))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": [
     "challenge"
    ]
   },
   "source": [
    "## Challenge\n",
    "\n",
    "Write a function `enough_cys` that will return a list of sequences with eight or more cysteine residues. (Hint: A cyteine residue is represented by a 'C' character).\n",
    "\n",
    "Once the function is written, we should be able to type:\n",
    "\n",
    "```python\n",
    "cys_seqs = enough_cys(seqs, 8)\n",
    "```\n",
    "\n",
    "and get a list of sequences with 8+ cysteine residues."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": [
     "solution"
    ]
   },
   "source": [
    "## Solution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 621,
   "metadata": {
    "scrolled": false,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": [
     "solution"
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total sequences: 129\n",
      "Sequences with >= 8 cyteines 129\n"
     ]
    }
   ],
   "source": [
    "def enough_cys(sequences, min_cys=8):\n",
    "    \"\"\"\n",
    "    Given a list of sequences (scikit-bio Sequence or strings), return a list\n",
    "    with `min_cys` or more cysteine residues.\n",
    "    \"\"\"\n",
    "    cys = []\n",
    "    for seq in sequences:\n",
    "        # https://docs.python.org/3.6/library/stdtypes.html#str.count\n",
    "        if seq.count('C') >= min_cys:\n",
    "            # print('>{seq_id}'.format(seq_id=seq_id))\n",
    "            # print(seq)\n",
    "            cys.append(seq)\n",
    "    \n",
    "    return cys\n",
    "\n",
    "cys_seqs = enough_cys(seqs, 8)\n",
    "\n",
    "print(\"Total sequences:\", len(seqs))\n",
    "print(\"Sequences with >= 8 cyteines\", len(cys_seqs))\n",
    "\n",
    "# Bonus: You can use `assert` with some condition to test expected outputs of your function\n",
    "# If any assertion fails, an error is raised. Otherwise they just execute silently.\n",
    "assert len(enough_cys([\"M\" + \"CA\"*10, \"MCACAC\"])) == 1\n",
    "assert len(enough_cys([\"M\" + \"CA\"*10, \"MCACAC\"], min_cys=3)) == 2\n",
    "assert \"MCACAC\" not in enough_cys([\"M\" + \"CA\"*10, \"MCACAC\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "We can also write a set of sequences back to a file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 603,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "with open('evasins_8cys.fasta', 'w') as outfile:\n",
    "    for seq in cys_seqs:\n",
    "        # The Sequence object from scikit-bio implements a 'write' method\n",
    "        # that writes the sequence in FASTA format to an open file handle\n",
    "        seq.write(outfile)\n",
    "\n",
    "# Let's reassign the variable 'seqs' to point to our set with 8+ cysteines\n",
    "seqs = cys_seqs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Plotting with Plotly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 605,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: plotly==2.2.1 in /usr/local/lib/python3.5/dist-packages\n",
      "Requirement already satisfied: six in /usr/local/lib/python3.5/dist-packages (from plotly==2.2.1)\n",
      "Requirement already satisfied: decorator>=4.0.6 in /usr/local/lib/python3.5/dist-packages (from plotly==2.2.1)\n",
      "Requirement already satisfied: nbformat>=4.2 in /usr/local/lib/python3.5/dist-packages (from plotly==2.2.1)\n",
      "Requirement already satisfied: requests in /usr/local/lib/python3.5/dist-packages (from plotly==2.2.1)\n",
      "Requirement already satisfied: pytz in /usr/local/lib/python3.5/dist-packages (from plotly==2.2.1)\n",
      "Requirement already satisfied: jsonschema!=2.5.0,>=2.4 in /usr/local/lib/python3.5/dist-packages (from nbformat>=4.2->plotly==2.2.1)\n",
      "Requirement already satisfied: traitlets>=4.1 in /usr/local/lib/python3.5/dist-packages (from nbformat>=4.2->plotly==2.2.1)\n",
      "Requirement already satisfied: ipython-genutils in /usr/local/lib/python3.5/dist-packages (from nbformat>=4.2->plotly==2.2.1)\n",
      "Requirement already satisfied: jupyter-core in /usr/local/lib/python3.5/dist-packages (from nbformat>=4.2->plotly==2.2.1)\n",
      "Requirement already satisfied: urllib3<1.23,>=1.21.1 in /usr/local/lib/python3.5/dist-packages (from requests->plotly==2.2.1)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.5/dist-packages (from requests->plotly==2.2.1)\n",
      "Requirement already satisfied: idna<2.7,>=2.5 in /usr/local/lib/python3.5/dist-packages (from requests->plotly==2.2.1)\n",
      "Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /usr/local/lib/python3.5/dist-packages (from requests->plotly==2.2.1)\n"
     ]
    }
   ],
   "source": [
    "!pip install plotly==2.2.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 606,
   "metadata": {
    "scrolled": false,
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [],
   "source": [
    "# By importing Plotly this way we can avoid 'logging in' to use it\n",
    "import plotly.graph_objs as go\n",
    "from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 607,
   "metadata": {
    "scrolled": false,
    "slideshow": {
     "slide_type": ""
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>requirejs.config({paths: { 'plotly': ['https://cdn.plot.ly/plotly-latest.min']},});if(!window.Plotly) {{require(['plotly'],function(plotly) {window.Plotly=plotly;});}}</script>"
      ],
      "text/vnd.plotly.v1+html": [
       "<script>requirejs.config({paths: { 'plotly': ['https://cdn.plot.ly/plotly-latest.min']},});if(!window.Plotly) {{require(['plotly'],function(plotly) {window.Plotly=plotly;});}}</script>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# We need to initialise offline plotly for the notebook so our charts display in-line\n",
    "init_notebook_mode(connected=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "### A histogram of sequence lengths\n",
    "\n",
    "Plotly histograms: https://plot.ly/python/histograms/\n",
    "\n",
    "We'd like to be able to visualize the distribution of sequence lengths in our evasin sequence set."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": [
     "challenge"
    ]
   },
   "source": [
    "## Challenge\n",
    "\n",
    "Write a **function** to return a list of lengths, given a list of sequences.\n",
    "\n",
    "eg, we want to be able to write:\n",
    "```\n",
    "lengths = get_lengths(sequences)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": [
     "solution"
    ]
   },
   "source": [
    "## Solution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 608,
   "metadata": {
    "tags": [
     "solution"
    ]
   },
   "outputs": [],
   "source": [
    "def get_lengths(sequences):\n",
    "    \"\"\"\n",
    "    Returns a list of sequence lengths, given a list of sequences.\n",
    "    \"\"\"\n",
    "    return [len(seq) for seq in sequences]\n",
    "\n",
    "\n",
    "# or without a list comprehension\n",
    "def get_lengths(sequences):\n",
    "    lengths = []\n",
    "    for seq in sequences:\n",
    "        lengths.append(len(seq))\n",
    "    return lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 609,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[146, 147, 122, 205, 146]"
      ]
     },
     "execution_count": 609,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lengths = get_lengths(seqs)\n",
    "lengths[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 610,
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [
    {
     "data": {
      "application/vnd.plotly.v1+json": {
       "data": [
        {
         "type": "histogram",
         "x": [
          146,
          147,
          122,
          205,
          146,
          179,
          145,
          148,
          141,
          118,
          104,
          136,
          145,
          205,
          206,
          107,
          96,
          130,
          124,
          177,
          147,
          127,
          103,
          107,
          148,
          131,
          100,
          100,
          132,
          120,
          205,
          150,
          160,
          202,
          126,
          148,
          118,
          117,
          215,
          123,
          171,
          135,
          172,
          123,
          136,
          163,
          136,
          130,
          147,
          148,
          114,
          149,
          115,
          153,
          145,
          133,
          143,
          205,
          154,
          137,
          118,
          145,
          145,
          152,
          206,
          155,
          107,
          160,
          127,
          141,
          121,
          117,
          136,
          163,
          121,
          119,
          109,
          127,
          204,
          151,
          119,
          150,
          150,
          136,
          114,
          1988,
          203,
          141,
          145,
          113,
          109,
          198,
          115,
          160,
          150,
          147,
          102,
          155,
          89,
          114,
          150,
          114,
          137,
          115,
          132,
          207,
          152,
          118,
          150,
          117,
          107,
          105,
          140,
          149,
          145,
          129,
          145,
          147,
          138,
          146,
          146,
          147,
          125,
          215,
          106,
          133,
          146,
          137,
          120
         ],
         "xbins": {
          "size": 1
         }
        }
       ],
       "layout": {}
      },
      "text/html": [
       "<div id=\"df572e50-6bab-4d0f-a3a2-4378bbacebb5\" style=\"height: 525px; width: 100%;\" class=\"plotly-graph-div\"></div><script type=\"text/javascript\">require([\"plotly\"], function(Plotly) { window.PLOTLYENV=window.PLOTLYENV || {};window.PLOTLYENV.BASE_URL=\"https://plot.ly\";Plotly.newPlot(\"df572e50-6bab-4d0f-a3a2-4378bbacebb5\", [{\"xbins\": {\"size\": 1}, \"x\": [146, 147, 122, 205, 146, 179, 145, 148, 141, 118, 104, 136, 145, 205, 206, 107, 96, 130, 124, 177, 147, 127, 103, 107, 148, 131, 100, 100, 132, 120, 205, 150, 160, 202, 126, 148, 118, 117, 215, 123, 171, 135, 172, 123, 136, 163, 136, 130, 147, 148, 114, 149, 115, 153, 145, 133, 143, 205, 154, 137, 118, 145, 145, 152, 206, 155, 107, 160, 127, 141, 121, 117, 136, 163, 121, 119, 109, 127, 204, 151, 119, 150, 150, 136, 114, 1988, 203, 141, 145, 113, 109, 198, 115, 160, 150, 147, 102, 155, 89, 114, 150, 114, 137, 115, 132, 207, 152, 118, 150, 117, 107, 105, 140, 149, 145, 129, 145, 147, 138, 146, 146, 147, 125, 215, 106, 133, 146, 137, 120], \"type\": \"histogram\"}], {}, {\"showLink\": true, \"linkText\": \"Export to plot.ly\"})});</script>"
      ],
      "text/vnd.plotly.v1+html": [
       "<div id=\"df572e50-6bab-4d0f-a3a2-4378bbacebb5\" style=\"height: 525px; width: 100%;\" class=\"plotly-graph-div\"></div><script type=\"text/javascript\">require([\"plotly\"], function(Plotly) { window.PLOTLYENV=window.PLOTLYENV || {};window.PLOTLYENV.BASE_URL=\"https://plot.ly\";Plotly.newPlot(\"df572e50-6bab-4d0f-a3a2-4378bbacebb5\", [{\"xbins\": {\"size\": 1}, \"x\": [146, 147, 122, 205, 146, 179, 145, 148, 141, 118, 104, 136, 145, 205, 206, 107, 96, 130, 124, 177, 147, 127, 103, 107, 148, 131, 100, 100, 132, 120, 205, 150, 160, 202, 126, 148, 118, 117, 215, 123, 171, 135, 172, 123, 136, 163, 136, 130, 147, 148, 114, 149, 115, 153, 145, 133, 143, 205, 154, 137, 118, 145, 145, 152, 206, 155, 107, 160, 127, 141, 121, 117, 136, 163, 121, 119, 109, 127, 204, 151, 119, 150, 150, 136, 114, 1988, 203, 141, 145, 113, 109, 198, 115, 160, 150, 147, 102, 155, 89, 114, 150, 114, 137, 115, 132, 207, 152, 118, 150, 117, 107, 105, 140, 149, 145, 129, 145, 147, 138, 146, 146, 147, 125, 215, 106, 133, 146, 137, 120], \"type\": \"histogram\"}], {}, {\"showLink\": true, \"linkText\": \"Export to plot.ly\"})});</script>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "bins = dict(size=1)\n",
    "# bins = dict(start=0.5, end=300.5, size=1)\n",
    "data = [go.Histogram(x=lengths, xbins=bins)]\n",
    "\n",
    "iplot(data, filename='Evasin sequence length distribution')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "The canonical evasins are ~120 - 140 residues long. Our set contains some sequences past 200 residues in length. This might warrant further investigation (but not today !).\n",
    "\n",
    "Let's remove all the sequences longer than 300 residues."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 611,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Uncharacterized protein OS=Phytomonas sp. isolate EM1 GN=GSEM1_T00005925001 PE=4 SV=1']\n"
     ]
    }
   ],
   "source": [
    "long_seqs = [seq.metadata['description'] for seq in seqs if len(seq) > 300]\n",
    "print(long_seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 612,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "128"
      ]
     },
     "execution_count": 612,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sane_seqs = [seq for seq in seqs if len(seq) <= 300]\n",
    "len(sane_seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 613,
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [
    {
     "data": {
      "application/vnd.plotly.v1+json": {
       "data": [
        {
         "type": "histogram",
         "x": [
          146,
          147,
          122,
          205,
          146,
          179,
          145,
          148,
          141,
          118,
          104,
          136,
          145,
          205,
          206,
          107,
          96,
          130,
          124,
          177,
          147,
          127,
          103,
          107,
          148,
          131,
          100,
          100,
          132,
          120,
          205,
          150,
          160,
          202,
          126,
          148,
          118,
          117,
          215,
          123,
          171,
          135,
          172,
          123,
          136,
          163,
          136,
          130,
          147,
          148,
          114,
          149,
          115,
          153,
          145,
          133,
          143,
          205,
          154,
          137,
          118,
          145,
          145,
          152,
          206,
          155,
          107,
          160,
          127,
          141,
          121,
          117,
          136,
          163,
          121,
          119,
          109,
          127,
          204,
          151,
          119,
          150,
          150,
          136,
          114,
          203,
          141,
          145,
          113,
          109,
          198,
          115,
          160,
          150,
          147,
          102,
          155,
          89,
          114,
          150,
          114,
          137,
          115,
          132,
          207,
          152,
          118,
          150,
          117,
          107,
          105,
          140,
          149,
          145,
          129,
          145,
          147,
          138,
          146,
          146,
          147,
          125,
          215,
          106,
          133,
          146,
          137,
          120
         ],
         "xbins": {
          "size": 1
         }
        }
       ],
       "layout": {}
      },
      "text/html": [
       "<div id=\"e2740ea7-8a34-44a1-b136-793a80077899\" style=\"height: 525px; width: 100%;\" class=\"plotly-graph-div\"></div><script type=\"text/javascript\">require([\"plotly\"], function(Plotly) { window.PLOTLYENV=window.PLOTLYENV || {};window.PLOTLYENV.BASE_URL=\"https://plot.ly\";Plotly.newPlot(\"e2740ea7-8a34-44a1-b136-793a80077899\", [{\"xbins\": {\"size\": 1}, \"x\": [146, 147, 122, 205, 146, 179, 145, 148, 141, 118, 104, 136, 145, 205, 206, 107, 96, 130, 124, 177, 147, 127, 103, 107, 148, 131, 100, 100, 132, 120, 205, 150, 160, 202, 126, 148, 118, 117, 215, 123, 171, 135, 172, 123, 136, 163, 136, 130, 147, 148, 114, 149, 115, 153, 145, 133, 143, 205, 154, 137, 118, 145, 145, 152, 206, 155, 107, 160, 127, 141, 121, 117, 136, 163, 121, 119, 109, 127, 204, 151, 119, 150, 150, 136, 114, 203, 141, 145, 113, 109, 198, 115, 160, 150, 147, 102, 155, 89, 114, 150, 114, 137, 115, 132, 207, 152, 118, 150, 117, 107, 105, 140, 149, 145, 129, 145, 147, 138, 146, 146, 147, 125, 215, 106, 133, 146, 137, 120], \"type\": \"histogram\"}], {}, {\"showLink\": true, \"linkText\": \"Export to plot.ly\"})});</script>"
      ],
      "text/vnd.plotly.v1+html": [
       "<div id=\"e2740ea7-8a34-44a1-b136-793a80077899\" style=\"height: 525px; width: 100%;\" class=\"plotly-graph-div\"></div><script type=\"text/javascript\">require([\"plotly\"], function(Plotly) { window.PLOTLYENV=window.PLOTLYENV || {};window.PLOTLYENV.BASE_URL=\"https://plot.ly\";Plotly.newPlot(\"e2740ea7-8a34-44a1-b136-793a80077899\", [{\"xbins\": {\"size\": 1}, \"x\": [146, 147, 122, 205, 146, 179, 145, 148, 141, 118, 104, 136, 145, 205, 206, 107, 96, 130, 124, 177, 147, 127, 103, 107, 148, 131, 100, 100, 132, 120, 205, 150, 160, 202, 126, 148, 118, 117, 215, 123, 171, 135, 172, 123, 136, 163, 136, 130, 147, 148, 114, 149, 115, 153, 145, 133, 143, 205, 154, 137, 118, 145, 145, 152, 206, 155, 107, 160, 127, 141, 121, 117, 136, 163, 121, 119, 109, 127, 204, 151, 119, 150, 150, 136, 114, 203, 141, 145, 113, 109, 198, 115, 160, 150, 147, 102, 155, 89, 114, 150, 114, 137, 115, 132, 207, 152, 118, 150, 117, 107, 105, 140, 149, 145, 129, 145, 147, 138, 146, 146, 147, 125, 215, 106, 133, 146, 137, 120], \"type\": \"histogram\"}], {}, {\"showLink\": true, \"linkText\": \"Export to plot.ly\"})});</script>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "bins = dict(size=1)\n",
    "data = [go.Histogram(x=get_lengths(sane_seqs), xbins=bins)]\n",
    "\n",
    "iplot(data, filename='Evasin sequence length distribution')"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
